KEP 5836: Add KEP for Scheduler Preemption for In-Place Pod Resize (alpha) (#5932)
kubernetes/enhancements: keps/sig-scheduling/5836-scheduler-preemption-for-ippr/README.md

request sitting there indefinitely.

#### Story 2: Reduction of disruption for critical workloads.

The Scheduler would preferentially evict lower-priority pods before an external controller, such as  VPA's 
[InPlaceOrRecreate mode](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler/enhancements/4016-in-place-updates-support#in-place-updates), 
would be forced to fall back to the more disruptive action of evicting and recreating the critical pod itself.

#### Story 3: Driving cluster autoscaling

Preempted pods managed by workload controllers are likely to make their way back to the scheduling queue. By moving the 
resource deficit from an indefinite Deferred status into the active scheduling queue, preempted pods will naturally trigger 
cluster autoscalers to provision new capacity. 

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

> No risks have been identified.

Reviewer: I definitely think there are risks here. Maybe you already implicitly assume some mitigations or have explanations why these are acceptable, but at least the two that I have on my mind are:

1. Performance - if we would have significant number of deferred pods, periodic reconsideration of those can visibly affect scheduling throughput.
Presumably the mitigation is triggering reconsideration of those only if something on the node changed.

2. Interaction with workload-aware-preemption - the resize may trigger preemption of potentially pretty large workload.
Arguably it's probably ok given it's lower priority, but it's worth mentioning here that the effect may be visibly larger than the on node itself.

as the primary goal is simply to trigger preemption. Instead of searching for a suitable node, the Scheduler 
will treat these resizes as a `FitError`, automatically moving them to the preemption phase. Within 
the preemption plugin, the search for victims must be restricted to the same node as the `Deferred` pod. Furthermore, the 
Scheduler needs to temporarily remove the `Deferred` pod from the node snapshot during calculations to ensure its current 
resource usage doesn't interfere with the preemption logic.

Taking all this into account, the logic for processing a `Deferred` resize is as follows:

1. **Identify Deferred Status**: Confirm the pod has a `Deferred` resize and is already bound to a node.
2. **Trigger Preemption**: Treat the request as a `FitError` to initiate existing Scheduler preemption logic.
3. **Isolate Node**: Within the Preemption plugin, narrow the victim search exclusively to the pod's node.
4. **Snapshot Adjustment**: Temporarily remove the `Deferred` pod from the node snapshot to calculate required space 
accurately.
5. **Calculate Victims**: Identify suitable preemption victims and then restore the pod to the snapshot.

> 6. **Update Status**: Report the success or failure of the preemption attempt in the pod status. If preemption is insufficient 

Author: I have thought a little more and I think we have to come back to @dom4ha's original point that reporting success may not be the end of the story. Even though the _scheduler_ does not need to do anything to keep blocking the newly requested resources, we may need to do something on the kubelet side to ensure the correct resize is retried.

The scenario I have in mind is multiple deferred resizes on the node. Say two pods Pod A and Pod B are both deferred; then scheduler preempts for Pod A. I need to confirm, but I believe there are scenarios where the kubelet may (with today's logic) preferentially choose to resize Pod B over Pod A. 

So we need either:
- some mechanism to communicate to the kubelet, when a preemption occurs, which deferred resize triggered it so that it can correctly reattempt it
- or some mechanism to ensure that the Pod A gets added back to the scheduling queue if the kubelet chooses to resize Pod B over Pod A.

I have some high level ideas of how to solve this, but I need to think more about which one is the best. Leaving this comment here to record the thought and will come back to address it. 

cc @tallclair

> 6. **Update Status**: Report the success or failure of the preemption attempt in the pod status. If preemption is insufficient 

Author: I looked into this a bit more. I think we can handle it the second way because this is how scheduling already works today.

IIUC, with scheduling, if a pod cannot be scheduled:
- it triggers preemption
- it sets the Unschedulable condition
- it gets moved to the Unschedulable queue
- when the preemption victim is removed, all unschedulable pods get moved back to the active (or backoff) queue

So with deferred resize, we can try something analogous:
- it triggers preemption
- it sets the ResizeUnschedulable condition
- it gets moved to the Unschedulable queue
- when the preemption victim is removed, all unschedulable pods get moved back to the active (or backoff) queue. If some other pod has taken the space in the meantime, it can trigger preemption again.

One thing though - there is a race condition; it's possible that the kubelet does not finish resizing either pod A or pod B before the scheduler reevaluates pod A. But I think what will happen during that reevaluation is that the scheduler will see that the resize now 'fits' on the node when doing its normal node fit check. In this case I think what we should do is to put the resize back into the Unschedulable queue. Then it will sit there until one of the following happens:
- pod A gets correctly resized and loses its deferred condition. Since scheduler is watching for this, it can then remove it from the Unschedulable queue
- pod B gets resized (even though scheduler preempted a victim for pod A). Now there is not enough room for pod A's resize, and it is correctly sitting in the Unschedulable queue
- pod A and pod B both get resized, and are removed from the scheduling queues

Hopefully this makes sense. I added a section to the KEP with a diagram to capture these thoughts.

> 6. **Update Status**: Report the success or failure of the preemption attempt in the pod status. If preemption is insufficient 

Reviewer 2: Keeping pods in Unschedulable queue make sense as kubelet is SoT, so it can always take a different decision than scheduler imagined, so we need to have a mechanism of eventual consistency.

The obvious concern is that waiting pods in unschedulable queue needs to be reevaluated whenever something changes in a cluster. It consumes scheduler cycles, so we need to make sure that only events on a given node can wake them up. I see it's in the KEP already in the Risks and Mitigations section.

We also need to make sure the preemption logic is consistent between scheduler and kubelet, otherwise it may lead to deadlocks (scheduler thinks podA is more important and preempted, but kubelet thinks podB is more important and does not resize podA). I think so far this is true and we need to make sure this Workload aware scheduling does not change it.

> 6. **Update Status**: Report the success or failure of the preemption attempt in the pod status. If preemption is insufficient 

Author: Marking as resolved; please reopen if there is something here I haven't addressed yet.

  * We would have to define the matrix of interactions between two independent preemption fields, leading to ambiguous states 
  (e.g. what happens if `preemptionPolicy` is set to `Never` but `resizePreemptionPolicy` is set to `PreemptLowerPriority`)?
  * Adds code complexity, including updates to the Pod API, validation logic, and new branches in the Scheduler.

If a pod is marked with `preemptionPolicy: PreemptLowerPriority`, the user has already communicated that this workload
is more important than lower-priority tasks. Whether that importance is manifested during initial placement or vertical 
scale-up, the intent of the priority remains the same.

### Pod Priority, Graceful Termination, and Pod Disruption Budget

Because we are reusing the existing Scheduling preemption logic, the behavior will be aligned with existing preemption
behavior for new pods. This means:

  * Pods will be preempted according to their Pod Priority.
  * The graceful termination period of victim pods will be honored.
  * PodDisruptionBudget is supported, but not guaranteed. 

### ResizeUnschedulable Pod Condition

> The Scheduler will own a new `ResizeUnschedulable` condition type in the pod status. This condition will be present only after

Author: > This is exactly how the initial scheduling works as well in case pods are unschedulable, so there should be no need to have any logic dedicated to the resizing.

Thank you! I took a quick look at the existing scheduling logic, and IIUC by having deferred resizes be treated as a `FitError`, this should allow us to reuse the same active queue / backoff queue / unschedulable pods management logic that is already implemented by the scheduler today, is that correct? I've added a section to the KEP to clarify this explicitly.

> I think there will be time to discuss details, but just to highlight, we should also take into consideration cases like when a higher priority pod needs to use resources reserved by the pod-during-resize. We probably don't want to kill the orignal pod if it's not necessary. 

I think I see the concern -- essentially the race between a `Deferred` resize and a new, higher-priority pod arriving. Is my understanding correct?

From the scheduler's view, once the spec is updated, the resources are already reserved as we discussed in  https://github.com/kubernetes/enhancements/pull/5932#discussion_r2864880440. I think that means that If a higher-priority pod comes in and the only way to fit it is by taking the space the Deferred pod is trying to grow into, the standard preemption logic applies. This might mean the resizing pod itself gets evicted if it’s the best victim candidate. I agree that we would rather not kill pods unnecessarily, but for the specific case that you are describing, this behavior seems consistent with the rest of the scheduler's logic. Perhaps I can add this to the known risks.

As a side note, this is why I have started conversations about introducing integration with node upsizing; it can prevent unnecessary disruptions such as this. 

> Scheduler also needs to keep track which pods is preempting what, so there are a few details we will have to have a closer look.

How is this handled with scheduler preemption today? I saw the `nominatedPods` map, so I wonder if that is the right way to track for resize as well, or do you think a separate mechanism is needed? I can see that conflating nominated pods with resize might be a bit confusing.

> The Scheduler will own a new `ResizeUnschedulable` condition type in the pod status. This condition will be present only after

Reviewer 2: > I think I see the concern -- essentially the race between a Deferred resize and a new, higher-priority pod arriving. Is my understanding correct?

Yes, that's the concern. Please add it to the known risks.

> How is this handled with scheduler preemption today? I saw the nominatedPods map, so I wonder if that is the right way to track for resize as well, or do you think a separate mechanism is needed

Since we don't need nominations, we don't need to worry about it anymore.

> The Scheduler will own a new `ResizeUnschedulable` condition type in the pod status. This condition will be present only after

Author: > Yes, that's the concern. Please add it to the known risks.

done.

3. Keep the preemption behavior for `Deferred` resizes aligned with the preemption behavior for newly created pods. 

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

1. Perform the entire scheduling cycle (including binding, DRA, etc) on `Deferred` resizes; the Scheduler should trigger 
preemption and nothing more. 
2. Coordinate with higher-level autoscalers such as VPA or CA on the preemption decision. This will be considered out
of scope for this KEP, but may be considered as a future enhancement. 

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

In-Place Pod Resize requests that result in a `Deferred` resize status will trigger scheduler preemption, keeping the 
behavior aligned with scheduler preemption for newly created pods. 

> The Scheduler will watch for `Deferred` resizes. To make room for the resize, it will try to preempt (evict) lower-priority pods to make the resize possible. The Kubelet today watches for pod removals, and will react to the eviction by reattempting the previously `Deferred` resize. 

Reviewer 2: It should watch also whether the requested resources doesn't change. If they do, it may require removing old and creating new deferred pod in the scheduling queue.

I also wonder what should happen on scheduler restart. I guess we should also put deferred pods to the scheduling queue for all pods for which the resize haven't succeeded yet. That means that in some situation we may send victim removal requests twice.

Reviewer 3: > It should watch also whether the requested resources doesn't change. If they do, it may require removing old and creating new deferred pod in the scheduling queue.

Why removing and re-adding the pod? I think it will be sufficient to just update the deferred pod in the queue with the new requests. It can trigger the requeueing then.

Reviewer 2: Sure, I ment to update it.

Author: > It should watch also whether the requested resources doesn't change. If they do, it may require removing old and creating new deferred pod in the scheduling queue.

Added to the kep that it should watch both for resource changes (which requires updating the pod in the scheduling queue, and moving it to the activeq if it is not already there); and also that it needs to watch if the deferred condition gets removed.

> I also wonder what should happen on scheduler restart. I guess we should also put deferred pods to the scheduling queue for all pods for which the resize haven't succeeded yet. That means that in some situation we may send victim removal requests twice.

I added some sections about this to the KEP. Yeah, I think putting all deferred pods in the scheduling queue on scheduler restart is the best answer. We could consider checkpointing or some other mechanism to make the preemption mechanism stateful, but I think the minor risk of sending victim removal requests twice is not worth that kind of complexity. I added sending additional victim removal requests to the known risks for this scenario as well.

It seems to me that this risk also exists with the current scheduler preemption logic today?

Reviewer 2: > We could consider checkpointing or some other mechanism to make the preemption mechanism stateful, but I think the minor risk of sending victim removal requests twice is not worth that kind of complexity

As I said in other comment https://github.com/kubernetes/enhancements/pull/5932#discussion_r3194617723, we should not send duplicated deletion requests (once deletionTimestamp is set), but we may potentially pick different victims. We need to discuss how big this concern is (to be addressed in alpha or beta).

> It seems to me that this risk also exists with the current scheduler preemption logic today?

Setting NNN in api-server prevents repeating the preemption process (the risk of picking different victims).

Author: Let's continue discussion of whether to set NNN in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3220939005. I've added the options to the body of the KEP, noting that decision about this is a beta-blocker.

information in the pod status. This would allow external controllers to provision additional capacity, if desired. This will 
be exposed as a new `ResizeUnschedulable` pod condition owned by the Scheduler, as described in more detail below. 

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1: Pods in the Deferred resize status no longer require manual interversion.

Users are enabled to instruct the system to automatically evict lower-priority workloads rather than the resize 
request sitting there indefinitely.

#### Story 2: Reduction of disruption for critical workloads.

The Scheduler would preferentially evict lower-priority pods before an external controller, such as  VPA's 
[InPlaceOrRecreate mode](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler/enhancements/4016-in-place-updates-support#in-place-updates), 

> would be forced to fall back to the more disruptive action of evicting and recreating the critical pod itself.

Reviewer 2: Does the timing of the preemption process matters in some way? Is the time for the fallback known (deadline for the operation)? What if it's shorter than the victim grace period? Maybe we don't have to think about it in alpha, but it may be something worth considering for beta.

Author: For VPA specifically, the time for the fallback is 5 minutes, which seems long enough to cover most scenarios. I believe they also have an open item to make that fallback time configurable

I'll add it as a beta blocker though, to consider these scenarios.

Author: I have question for you though. If the victim's grace period is very long, and there is some other activity on the node (such as pod deletion) that results in the deferred pod being moved back into the 'activeQ' before the victim is terminated, would that result in a second victim being preempted? 

I wonder if this is a case for having to keep track of which pod is preempting what via nominations (as you mentioned in https://github.com/kubernetes/enhancements/pull/5932#discussion_r2873470257) so that we don't retry preemption while the initial victim pod is still terminating. WDYT?

Reviewer 2: It's a good question. Scheduler has two mechanism preventing double preemption that should work together:

1. Skipping preemption entirely if the pod has nomination, it is resolvable and if there is still an eligible pod that is in grace period:
https://github.com/kubernetes/kubernetes/blob/8d4ef007d33716a96701aef51734ad311dedad52/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go#L370
2. If the previous check passes (for instance if we haven't had nomination), scheduler also skips sending victim delete requests if they are already in the grace period (have deletionTimestamp set):
https://github.com/kubernetes/kubernetes/blob/8d4ef007d33716a96701aef51734ad311dedad52/pkg/scheduler/framework/preemption/executor.go#L211

So if we hadn't set nomination after the first preemption attempt, we would skip step 1 and run preemption process again. As long as nothing has changed in the meantime, it should produce the same set of victims and will either send duplicated deletion requests or notice that victims are already in grace period. But in case something had changed, the preemption would pick different set of victims and perform unnecessary preemption. 

That means that setting nomination is something that we indeed should consider, even thought we don't need its other properties (pod is already on a node).

So the concern is that we'd semantically change the meaning of NNN, since we don\t used it for existing pods (if we decide to persist it in the api-server), which potentially may have a few implications (for instance for autoscalers)

@macsko @sanposhiho @wojtek-t @Argh4k Is it worth to use NNN to prevent potential unnecessary preemptions and not generate load on repeating preemption? I'd say that it's not a problem for alpha, but we should address it somehow for beta.

Author: > So the concern is that we'd semantically change the meaning of NNN

What if, for deferred pods, we use something like NNN to prevent double preemption, but we don't surface it in any API field / don't persist it to the api-server - just track it internally in some struct? That leaves us flexibility to resolve the issue in this way but allow us to change it later as it would be an unsurfaced implementation detail.

I think better to avoid semantically changing the meaning of an already-exposed API field (NNN).

We would lose the information in the case of scheduler restart, but that as you've said in [your other comment](https://github.com/kubernetes/enhancements/pull/5932#discussion_r3188936434) should be quite rare?

I suggest we do this for alpha, and revisit the decision for beta.

Reviewer 2: That would be the middle ground option.

I agree with all you said, but the change in the scheduler itself would not be small either, so I'd rather start with not using NNN for alpha and simply tolerate a possible unnecessary preemption.

It shouldn't be that common either. Note that we clear NNN in some cases, so we also agree on suboptimal preemption behavior in certain cases (when a higher priority pod picks a space reserved by a lower priority nominations)

Author: Not handling for alpha and deferring decision to beta SGTM.

Reviewer 4: >  If the victim's grace period is very long, and there is some other activity on the node (such as pod deletion) that results in the deferred pod being moved back into the 'activeQ' before the victim is terminated, would that result in a second victim being preempted?

Shouldn't that be blocked by PreEnqueue of default preemption that checks whether a given pod is running any preemptions? https://github.com/kubernetes/kubernetes/blob/7988c94bf6fc9cfd3db9fb6c591219c236118bbe/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go#L183

Reviewer 2: Not by PreEnqueue, because we still want to schedule a pod-waiting-for-preemption if there is some other place feed up in the meantime. So we need to repeat filtering on every cluster change.

This code you linked prevents only rescheduling a pod when there is an async preemption in progress (scheduler haven't finished sending api requests).

Reviewer 3: > Is it worth to use NNN to prevent potential unnecessary preemptions and not generate load on repeating preemption?

I think it's worth to use NNN, but at the same time we have to be cautious how to handle them. In particular, we don't want to put such pods to nominator, so it may require additional code to handle this.

> What if, for deferred pods, we use something like NNN to prevent double preemption, but we don't surface it in any API field / don't persist it to the api-server - just track it internally in some struct?

In case of scheduler's restart we would lose this information.

Author: > In case of scheduler's restart we would lose this information.

Yeah, I noted this in my comment above. I wasn't sure if strictly avoiding double preemption in the case of scheduler restart was worth changing the semantics of the user-facing NNN field, particularly because it sounds like it would be very rare.

But I will defer to the opinion of scheduling leads on this if you all think it is worth it as you know better than me what the consequences are.

I noted all the options in the body of the KEP and have added it as a beta blocker to revisit the alpha implementation of not handling it. Proceeding with NNN appears to be the recommendation, so we can flesh out those details as we prepare for beta in the next release?

Reviewer 3: > I noted all the options in the body of the KEP and have added it as a beta blocker to revisit the alpha implementation of not handling it. Proceeding with NNN appears to be the recommendation, so we can flesh out those details as we prepare for beta in the next release?

Agree, let's defer the potential support of NNN to beta promotion discussion. In alpha it's fine to re-compute the preemption in worst case.

Reviewer 3: Btw, we could also just add a check to the DefaultPreemption for NodeName similar to NominatedNodeName. Then, we wouldn't need to set the NNN at all
https://github.com/kubernetes/kubernetes/blob/8d4ef007d33716a96701aef51734ad311dedad52/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go#L370

Author: > Btw, we could also just add a check to the DefaultPreemption for NodeName similar to NominatedNodeName. 

I'll need to think about this a bit more, but at first glance I think that would work. I added it as an additional option to consider. Thanks!

Reviewer 2: Not sure how that would help. How would we distinguish the first PostFilter run, which should trigger preemption, from the consecutive ones which shouldn't?

Reviewer 2: Anyway, let's not block on this. We can change the decision during implementation or defer it for beta.

Reviewer: > Not sure how that would help. How would we distinguish the first PostFilter run, which should trigger preemption, from the consecutive ones which shouldn't?

If the pod has a nomination and the terminating victims won't let it run anyway, we need to run preemption anyway.
We would just build on that - if it's nominated we check if "once all terminating pods are removed, will there be enough space". If not - we preempt, if yes, we wait...

I'm definitely +1 to Maciek's idea - it sounds order of magnitude simpler than anything else.

Reviewer 2: I see, so we can trade not running unnecessary preemption attempts for the cost of delaying the first preemption until any overlapping preemption (two pods share the same victim) on this node is finished.

Anyway, it's a good idea until we use `max(desired, allocated, actuation)` resources for preemptor and victims, so slower preemptions will help avoiding unnecessary preemptions.

But when we start ignoring `desired` for victims and make the preemption process idempotent (we have an independent discussion about it), then performing preemption attempts more often should not be harmful, so we could trigger them earlier (reverse the tradeoff).

If there are a significant number of deferred pods, periodic processing of those pods can affect
scheduling throughput. To mitigate this, deferred pods will only be reconsidered if something
on the node has changed that could make preemption succeed.

#### Interaction with workload-aware preemption

With workload-aware preemption, the resize may end up triggering preemption of a potentially large
workload. In these cases, the preempted workload would always be lower priority than the pod 
that is being resized, so we consider this working as intended.

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### How Deferred Resizes will integrate into the Scheduling Queue

In the [UpdatePod event handler](https://github.com/kubernetes/kubernetes/blob/60433d43cf0bb83a2ac7d5e767137b3d510026ec/pkg/scheduler/eventhandlers.go#L147) that the Scheduler
already has, the Scheduler will check to see if a pod has recently been marked

> as having a `Deferred` resize. If so, the pod will be added to the scheduling queue.

Reviewer 3: That will complicate the logic in the scheduler, because now the pod will be in two places at the same time: cache (in node info) and scheduling queue. Any further pod update would require updating these two places, what needs a special handling.

We need to be careful with that, as the scheduler can take some assumptions that the pod lies in a single place.

Author: Thanks for bringing this up. I thought more and wanted to run this idea by you to see if it helps.

Rather than adding the pod snapshot to the scheduling queue directly, in the case of deferred resize, we can instead add a pointer to the pod in the cache. That would prevent the pod from being in two places. This can be implemented by changing the Pod stored in `PodInfo` to instead be an interface with a method `GetPod` to return the pod.
- For existing scenarios (i.e. scheduling new pods), the implementation of `GetPod` can simply return the pod that is stored. The goal is to have identical logic and behavior as today.
- For deferred resize scenarios, the implementation of `GetPod` can return the pod from the pod cache directly. This would mean we no longer have to store the pod separately, and updates to the pod in the cache will automatically propagate to the scheduling queue. We may need to 'snapshot' the deferred pod right before the binding / preemption cycle, to ensure it's not changing throughout.

This doesn't solve all the "double counting" issues though; we would still need to temporarily remove the pod from the node snapshot when making preemption calculations. I'm not sure if there are other places where we also have to specially handle this kind of thing.

Alternative solutions could look like one of these:
- We add the deferred pod itself to the scheduling queue, but do not use it as the source of truth. Instead, we look at the cache before we use deferred pods from the scheduling queue.
- We attempt to update the pod in both places whenever necessary and manually handle all cases where we might need to avoid “double counting” or other issues.

These alternatives I think could be more error-prone though.

What do you think of the idea? If it helps to understand it better, I pushed the interface change in the [first commit ](https://github.com/kubernetes/kubernetes/pull/137206/changes/fd37d615883e74689bc60a2c3bf6eea33da024ae)of the [PoC](pull/137206) to better illustrate what it could look like. The [second commit](https://github.com/kubernetes/kubernetes/pull/137206/changes/fa6253d32329382d137d328c6984ec9774e24dd7) shows how it would be used to handle deferred resizes.

cc @dom4ha to weigh in as well

Reviewer 2: @macsko is making changes to the scheduling queue, so I'd wait for his input and suggestions.

Reviewer 3: > Rather than adding the pod snapshot to the scheduling queue directly, in the case of deferred resize, we can instead add a pointer to the pod in the cache.

I'm worried that it might skip some relevant code paths, making the code even harder to manager. For example, PriorityQueue.Update won't be called, what might prevent the pod preemption from being retried (when the update is a relevant event). I feel like we should keep the pod object in two places, but ensure there are no cases that assume the pod to be in one place only.

Author: >  I feel like we should keep the pod object in two places, but ensure there are no cases that assume the pod to be in one place only.

SGTM. Updated the KEP accordingly.

The scheduling queue will now be shared between new pods and `Deferred` pods, using existing queuing 
logic to sort them by their current scheduling priority.

### Processing Deferred Resizes in the Scheduling Queue

When processing `Deferred` resizes, the Scheduler will skip most of the standard scheduling cycle, 
as the primary goal is simply to trigger preemption. Instead of searching for a suitable node, the Scheduler 

> will treat these resizes as a `FitError`, automatically moving them to the preemption phase. 

Reviewer 3: Does it mean that when the pod is in Deferred state, it cannot self-heal? For example when some pods disappear from the node in the meantime, making a sufficient space for the resized pod?

Author: The kubelet does react to pods disappearing (from removal, deletion, eviction) and perform the resize. The deferred resize will stay indefinitely only if there is nothing in the system that actively removes or downsizes other pods.

Reviewer 3: So how will scheduler react to such situation where trying the preemption is no longer necessary?

Author: The `Deferred` condition on the pod would disappear. So for that case perhaps scheduler should also watch for a pod moving out of the `Deferred` state and if so, remove it from the queue. WDYT?

Author: updated the kep to say that when the deferred condition is removed, we will remove it from the scheduling queues

> will treat these resizes as a `FitError`, automatically moving them to the preemption phase. 

Reviewer 3: It still requires calling at least PreFilter to populate the CycleState correctly. I'm not sure whether it should even run Filters, as some out-of-tree plugins might rely on them being run before the preemption.

Reviewer 3: Maybe we could just pass this pod through the scheduling process, but restricting it to its node? It may be simpler to implement

Author: Ack, yeah the state not being populated correctly was an issue in my existing PoC. I can try this approach and see if it's simpler.

Reviewer 2: Yes, it should rather follow the regular scheduling process, as running PreFilters may be necessary before Preemption in PostFilter kicks in. The only difference is that Filtering will be scoped to the specific node.

If it passes (because something has changed in the meantime), the pod may be simply removed for the queue and forgotten. If it fails, it will move to the PostFilter phase.

Author: updated both the kep and the poc to take the suggestion of following the standard scheduling process (limited to the node that it is already bound to)

Within the preemption plugin, the search for victims must be restricted to the same node as the `Deferred` pod. 
Due to workload-aware preemption, though the space needs to be freed up on the same node as the `Deferred` pod,

> this can result in preemption of pods on other nodes. 

Reviewer 3: How will it work when the pod belongs to a pod group and the pod group-level preemption is calculated?Will it be easy to adjust that code to cover this scenario? @dom4ha do you have any thoughts?

Reviewer 2: Moving my comment from other place here:

According to the WAS cycle, the PodGroup itself should get into scheduling queue, not the pod itself. Also the PodGroup should be going into preemption process, but with only one pod supposing to trigger preemption.

I think we should take this case out-of-scope for alpha (put it as beta blocker) and still handle PodGroup pods individually. One of the reasons is that currently PodGroup preemption disables preemption in PostFilter for individual pods and preemption is always global (cluster scoped). 

I'm working on a proposal to fix that (allow preemption PostFilter in the nested sub-scheduling cycles). We may proceed with some parts of it in 1.37, but it's not a change critical to this release, so it may be not fully ready.

Author: To clarify, are we discussing the scenario where a pod within a pod group is being upsized and needs to preempt? 

I think likely we would want to modify the WAS cycle in a similar way as we have elsewhere; resizing pods need to be locked to their current nodes while the WAS algorithm runs its feasibility checks. We'd also need to handle the case if multiple pods within a pod group are being resized. 

I'm happy to add it as a beta requirement until there is more clarity around the effort though, leaving it out of scope for alpha

Author: Added a section at the end of the kep to capture this: https://github.com/natasha41575/enhancements/blob/dccdad9db779786768167b4a3c597a13c43638e9/keps/sig-scheduling/5836-scheduler-preemption-for-ippr/README.md#interaction-with-workload-aware-preemption-1

  (e.g. what happens if `preemptionPolicy` is set to `Never` but `resizePreemptionPolicy` is set to `PreemptLowerPriority`)?
  * Adds code complexity, including updates to the Pod API, validation logic, and new branches in the Scheduler.

If a pod is marked with `preemptionPolicy: PreemptLowerPriority`, the user has already communicated that this workload
is more important than lower-priority tasks. Whether that importance is manifested during initial placement or vertical 
scale-up, the intent of the priority remains the same.

### Pod Priority, Graceful Termination, and Pod Disruption Budget

Because we are reusing the existing Scheduling preemption logic, the behavior will be aligned with existing preemption
behavior for new pods. This means:

  * Pods will be preempted according to their Pod Priority.
  * The graceful termination period of victim pods will be honored.
  * PodDisruptionBudget is supported, but not guaranteed. 

### Failures and Reconsideration of Deferred pods

#### ResizeUnschedulable Pod Condition

The Scheduler will own a new `ResizeUnschedulable` condition type in the pod status. This condition will be present only after

> a preemption attempt fails. For example: 

Reviewer 3: Will that be set by the core scheduler (in handleSchedulingFailure)?

Author: Yes, that was the idea

Author: clarified in-kep

    lastTransitionTime: "2026-02-23T15:23:13Z"
```

The `reason` will be set to `PreemptionFailed` and the `message` will provide more
details about the failure.

#### Handling retries

To ensure efficiency and avoid unnecessary reconciliation loops, the Scheduler will treat `Deferred` resizes that fail preemption as a 
standard `FitError` (Unschedulable). By classifying the resize failure this way, the existing Scheduler retry and backoff 
logic is triggered automatically, requiring no new dedicated retry infrastructure.

1.  **Queue Placement**: When a preemption attempt fails to find enough victims on the node to satisfy the resize, the pod is moved to 
the `unschedulablePods` pool within the scheduling queue.
2.  **Triggering Re-entry**: Because the pod is in the `unschedulablePods` pool, it will be moved back to the `activeQ` (or 
`backoffQ`) automatically when cluster events occur that could potentially change the outcome of the preemption logic. These events 
include:
    * **Pod Deletion**: A pod on the same node is deleted or removed (freeing up capacity).
    * **Pod Update**: A pod on the same node has its priority lowered or its resource requests decreased.

>     * **Node Update**: The node's `Allocatable` resources increase (e.g., via a Kubelet config update).

Reviewer 3: FitError has UnschedulablePlugins field that tells which plugins rejected this pod, and then scheduling queue evaluates the incoming events based on these plugins (using queueing hints). To have the UnschedulablePlugins populated, we have to run the scheduling algorithm, including filter plugins as I mentioned in another comment.

marking a significant shift in how Kubernetes operates. This is especially powerful when In-Place Pod Resize is integrated
with higher-level autoscaling controllers such as VPA's new `InPlaceOrRecreate` mode. However, users currently lack a 
way to control scale-up behavior when a node lacks the capacity to fulfill the request. Consequently, workloads may face 
disruptions such as being moved to a larger node, or suffering an OOM-kill because memory could not be scaled up in-place.

Scheduler preemption eliminates the gap by introducing an configurable ability to free up capacity on a fully-utilized
node to allow the scale up to succeed in-place.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

1. Introduce the ability to instruct the Scheduler to preempt lower-priority pods to make room for a `Deferred` resize.
2. Introduce a new Scheduler-owned condition in the pod status to communicate when preemption is insufficient to free up capacity.
3. Keep the preemption behavior for `Deferred` resizes aligned with the preemption behavior for newly created pods. 

> ### Non-Goals

Reviewer 2: I'd also add that we're not trying to change the fact that kubelet is SoT and that it can take arbitrary decisions without looping in scheduler.

Author: added!

scheduling throughput. To mitigate this, deferred pods will only be reconsidered if something
on the node has changed that could make preemption succeed.

#### Interaction with workload-aware preemption

With workload-aware preemption, the resize may end up triggering preemption of a potentially large
workload. In these cases, the preempted workload would always be lower priority than the pod 
that is being resized, so we consider this working as intended.

#### Race between a Deferred resize and a new higher-priority pod

From the scheduler's view, once the spec is updated, the resources are already reserved. It doesn't matter from the scheduler perspective whether the resize has been actuated by the Kubelet yet.

This means that if a new, higher-priority pod comes in and the only way to fit it is by taking the space the Deferred pod is trying to grow into, the standard preemption logic applies. This might mean the resizing pod itself gets evicted if it’s the best victim candidate. While we would rather not kill pods unnecessarily, this behavior is consistent with the rest of the scheduler's logic.

> #### Shifting Preemption Victims during Scheduler Restart

Reviewer 3: NominatedNodeName should prevent this. As long as the victims are pending deletion on the nominated node, no additional preemptions will be attempted

Author: Ack, added a note that using NNN (persisted to the api server) would prevent this.
Per discussion in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3220939005, I added the different options to the KEP and noted that finalizing the decision for handling this would be a beta blocker.

If a `Deferred` resize was mid-preemption when the scheduler crashed or restarted, the new scheduler instance might select a different victim 
pod than the original instance did. This can happen due to:

*   Changes in cluster state (new pods, node updates) during the scheduler's downtime.
*   Non-deterministic tie-breaking when multiple low-priority pods satisfy the resource requirement equally well.

In specific edge cases, this leads to redundant preemption. "Victim A" (targeted by the first scheduler) and "Victim B" (targeted by the second scheduler) may both be terminated to satisfy a single resize request.

Mitigations:

> *   **Idempotency:** The `Delete` API is idempotent. If the scheduler picks the same victim upon restart, the API server simply acknowledges the 

Reviewer 2: It should be already idempotent

Author: I'll rephrase to clarify it is already idempotent

request without further disruption.

> *   **Acceptable Waste:** In the Kubernetes priority model, ensuring the success of a higher-priority workload (the resizing pod) justifies the 

Reviewer 2: This would be really rare case, as Pods that require grace shutdown period are marked with `deletionTimestamp`, so should be spotted by new scheduler as well.

Author: There is more discussion around this in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3220939005, so will mark this thread as resolved and we can continue discussing there

potential loss of multiple lower-priority victims during a rare control-plane failure.

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### How Deferred Resizes will integrate into the Scheduling Queue

> In the [UpdatePod event handler](https://github.com/kubernetes/kubernetes/blob/60433d43cf0bb83a2ac7d5e767137b3d510026ec/pkg/scheduler/eventhandlers.go#L147) that the Scheduler already has, the Scheduler will check to see if a pod has recently been marked as having a `Deferred` resize. If so, the pod will be added to the scheduling queue. 

Reviewer 3: And in the AddPod handler in case the scheduler was restarted

Author: added this, thanks!

The scheduling queue will be shared between new pods and `Deferred` pods, using existing queuing logic to sort them by their current scheduling priority. 

The `UpdatePod` event handler will also watch for:

> * **If the pod's resources changed from last time**: We need to update our view of the pod, and move it to the active queue if it is not there already.

Reviewer 3: It may be tricky to implement to move the pod to the activeQ from within the event handler based on a specific event.

Author: Let's discuss further in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3221058391

Author: It looks like [schedulingqueue.Activate](https://github.com/kubernetes/kubernetes/blob/d8dc1e25b01e08916652fb89f600319f95571937/pkg/scheduler/backend/queue/scheduling_queue.go#L101-L106) already exists that allows us to move the pod to the active queue from wherever it is. 

I pushed a commit https://github.com/kubernetes/kubernetes/commit/47b549db42a614b5df98260efbf66c08faa049b7 to show what this looks like. Let me know if you see any issues with this.

implemented by changing the Pod stored in PodInfo to instead be an interface with a method GetPod to return the pod:

```go
type PodGetter interface {
    GetPod() *v1.Pod
}
```

* **For existing scenarios** (i.e. scheduling new pods): The implementation of `GetPod()` can simply return the pod that is stored. The goal is to have identical logic 
and behavior as today.
* **For deferred resize scenarios**: The implementation of `GetPod()` can return the pod from the pod cache directly. This would mean we no longer have to store the pod 
separately, and updates to the pod in the cache will automatically propagate to the scheduling queue. We may need to 'snapshot' the deferred pod right before the 
binding / preemption cycle, to ensure it's not changing throughout.

While the `PodGetter` interface provides a live reference to the cache, the Scheduler will perform a local snapshot/clone of the Pod object at the start of its individual scheduling cycle. This ensures that even if the Kubelet or another controller updates the Pod status in the cache during the Preemption calculation, the Scheduler operates on a consistent, immutable view of the resize request for that cycle.

> This doesn't solve all the "double counting" issues though; we would still need to temporarily remove the pod from the node snapshot when making preemption calculations

Reviewer 2: Do we need to "temporarily remove" pod from the snapshot? When we try to place the resized pod to the node it is already running, we would simply take max of existing and new resources.

Reviewer 4: I think that we do not it at all. In default preemption we create a snapshot of the node which we are checking for preemption and work on that: https://github.com/kubernetes/kubernetes/blob/7988c94bf6fc9cfd3db9fb6c591219c236118bbe/pkg/scheduler/framework/preemption/preemption.go#L420-L424

Author: The goal was to prevent double counting. The node resource fit plugin will add the resources of the pod being processed + node allocated resources. For a resizing pod, this means it gets counted twice (because it is the pod being processed, and because its resources are already allocated on the node).

But I think taking max of existing and new resources is a reasonable alternative and to be honest probably easier to implement. Node allocated resources is already max(desired, allocated, actual) for each pod, so what we can do is just simply modify the resource fit plugin to just use the node allocated resources (and not add the resizing pod's additional resources on top of it).

If node allocated resources is over 100% (which will always be the case for Deferred resizes), then we proceed with preemption. I've updated the kep accordingly. If the proposal looks reasonable to you, I'll mark this conversation as resolved.

as described in the section below.

*See alternative considerations for how to handle the scheduler having two places to find the deferred pod [here](#avoid-conflicting-sources-of-truth-for-the-pod-1).*

> ### Processing Deferred Resizes in the Scheduling Queue

Reviewer 3: How will the deferred resize handling look like for pods belonging to a pod group? Such pods follow a PodGroup scheduling cycle and then execute workload-aware preemption if needed

Author: This was brought up in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3095535915, and the suggestion was to take this out of scope for alpha but put it as a beta-blocker. WDYT?

Author: I added a section at the end of the kep to cover this: https://github.com/natasha41575/enhancements/blob/82ea5804c29cd5bbdaf21099e92785edde6a3c7e/keps/sig-scheduling/5836-scheduler-preemption-for-ippr/README.md#interaction-with-workload-aware-preemption-1

Reviewer 3: In that section you mention a case where the victim belongs to a pod group. We also have more tricky case where the deferred pod can belong to a pod group. Handling of that case won't be trivial, because pod group members are not put directly into the scheduling queue, but though a PodGroup abstraction. IIUC, for deferred pod group members we would still like to put them individually in the queue. This would require adjustments in the queue logic, not sure how large.

It would be great if you mention that in the KEP. Depending on the potential size of changes, we could defer such support to beta with a decision what to do with deferred pod group members for alpha.

Author: The section I linked did have a subsection mentioning if the deferred pod is part of the pod group. But the heading was the same as another subheading under risks so possible something got mixed up. Anyhow, I have updated it slightly and retitled it, here is the updated link: https://github.com/natasha41575/enhancements/blob/0a5ec2caffb775e7d6798d4e7dcf4bb2d9b52b54/keps/sig-scheduling/5836-scheduler-preemption-for-ippr/README.md#scope-of-interaction-with-workload-aware-preemption

Copying what it states now:

```
### Scope of Interaction with Workload-Aware Preemption

With workload-aware preemption, there are two scenarios to consider:

1. **The selected victim pod is part of a workload**:
   The scheduler preempts the entire workload. Since the preempted workload's priority is strictly lower than that of the resizing pod, this behaves as designed.

2. **The resizing pod itself is part of a workload**:
   * **High-Level Direction**: When a resizing pod belongs to a workload group (e.g., a PodGroup scheduled atomically), resize-induced preemption must evaluate the impact on the entire workload group. The scheduler must verify if the resize violates any workload-wide invariants (such as minimum member availability or atomic group scheduling rules).
   * **Queueing Behavior for Group Members**: A running pod that requests a resize must be queued and retried individually. If the pod belongs to a `PodGroup`, the scheduling queue instead treats the deferred pod as an independent individual pod. In Alpha, this will be implemented by having `isPodGroupMember` return `false` for deferred resize pods, allowing them to follow the standard individual queueing and backoff pathways.
   * **Alpha Scope**: The resizing pod is evaluated individually for preemption victim selection on its assigned node. The scheduler does not proactively trigger group-wide rescheduling or preemption of other members of the workload group.
   * **Beta Graduation**: Co-existence mechanics, including group-wide coordinated preemption (e.g., preempting other members of the same workload to balance resource usage or preventing preemption if the workload's group-wide health is already degraded), will be fully designed and finalized prior to Beta.
```

So tl;dr the KEP currently is taking @dom4ha's previous suggestion that for alpha it will be evaluated individually, but we will properly design the behavior for beta.

> for deferred pod group members we would still like to put them individually in the queue. This would require adjustments in the queue logic, not sure how large.

I think for alpha this can be implemented by having a special case in [isPodGroupMember](https://github.com/kubernetes/kubernetes/blob/8437feaf44b6978f6323b3165e14a3d67386f562/pkg/scheduler/backend/queue/scheduling_queue.go#L485) to return false for deferred pods, which will cause it to be queued as an indvidual pod. We can revisit this implementation as part of the design of coordinated preemption (in beta).

Reviewer 3: Sounds good, thanks

When processing `Deferred` resizes, the Scheduler will still perform a standard scheduling cycle. It will perform the normal scheduling and 

> binding evaluation, but will restrict its node evaluation search exclusively to the node the deferred pod is already running on. 

Reviewer 5: by the node name plugin?

Reviewer 5: If Yes, we should implement PreFilter and PreFilterResult on the node name plugin so that the deferred pod evaluation would be faster

Reviewer 2: We could handle it on the framework side and make the scheduling/preemption to be restricted to a single node.

I think it won't be regular scheduling anyway, as we don't want to apply scheduling time constraints like pod affinities, so it may be rather similar to the reprieve process.

Reviewer 5: I do not want to make the framework side dirty unnecessarily to treat these deferred pods. I believe having a prefilter ext point on the nodename plugin is much nicer in that sense. That also makes the preemption decision correct without any other "framework" side change at the same time (the deferred pod preemption can't pick up any other nodes)

Reviewer 2: Handling Pod resizes (even just the preemption side) requires deeper framework changes anyway. I agree we need to minimize them and defer if they are not strictly necessary in the first implementation, so let's to enumerate the reasons first.

- We definitely need to distinguish new Pods in the queue from the Pods-to-update.
- We need to have a way to disable some plugins to not check most of the scheduling constraints (like pod affinities)
- We need to avoid plugins from running Filter and PostFilter on all nodes, as the scope is a single node.

So we either need to
1. do it on the framework side - have the mechanisms that @natasha41575 described (plugins opt-outs) and what I proposed (limit the scope to a single node on the framework side) or
2. have specialized path - not follow strictly regular scheduling cycle (ScheduleOne), but start from defining resize specific cycle and implement ResizeOne function that would call only specific plugins.

I'd rather make the implementation generic as long as we are aware of all the consequences, but I'm not sure yet if we really are, so I'm open for the alternative.

Author: The proposal at the moment is closer to @dom4ha's option (1) rather than (2) because I thought it better to reuse existing scheduler code as much as possible rather than adding a completely new divergent path, as this is aligned with the goal of keeping preemption behavior for resize identical to preemption behavior for scheduling. The plugin opt-out / `ResizeInterestedPlugin` interface is specifically intended to ensure we don't break any plugins (in-tree or out-of-tree) that may not handle resize correctly. If this solution works, it is also more flexible and extensible (it allows other plugins the option to handle resize if they want to). 

Implementing a separate resize-specific cycle / `ResizeOne` function seems to potentially result in duplicating significant portions of complicated code?

Reviewer 5: The framework level change is surely necessary around the queue and cache because they are assuming the pods to queue are only unschedulable pods. But, not sure about the other parts,

> We need to have a way to disable some plugins to not check most of the scheduling constraints (like pod affinities)
> We need to avoid plugins from running Filter and PostFilter on all nodes, as the scope is a single node.

I am saying that if PreFilterResult is returned from the nodename plugin, all the other filtering would be skipped, right? Like we handle the node selector from daemonset pods.

---

As @natasha41575 also mentioned, I definitely want to avoid having a dedicated ResizeOne func separate from ScheduleOne completely. So, I am on (1) side too, but I don't like `what I proposed (limit the scope to a single node on the framework side)`. Because the nodename plugin can filter out any other nodes except pod.spec.nodename. My base thinking is like we should abstract what need to be abstracted so that ScheduleOne can handle deferred pods as well without any unnecessary framework side changes (I am not saying we can completely avoid framework side changes, but saying we should avoid as much as possible by packing deferred pod logic into plugins as much as possible, rather than leaking them to the framework side).

Author: > Because the nodename plugin can filter out any other nodes except pod.spec.nodename. My base thinking is like we should abstract what need to be abstracted so that ScheduleOne can handle deferred pods as well without any unnecessary framework side changes

The current PoC modifies `ScheduleOne` directly to overwrite the 'allNodes' struct to just be a single node. This happens in more than one place. I see @sanposhiho's point that this looks messy and using a plugin to filter it out prior to the scheduling cycle could be simpler. I want to take time to look at what the implementation would look like before forming my opinion, I can share the commit here when I have it (this has to wait until after my OOO though)

Reviewer 3: > I am saying that if PreFilterResult is returned from the nodename plugin, all the other filtering would be skipped, right?

Not really. PreFilter always has to be executed with all nodes, because it has to compute correct (anti-)affinity and topology spreading counts. Even if the NodeName plugin would return a single node in PreFilter, other plugins will execute as well, just the Filter will run on that one node.

Moreover, there are some constraints enforced at scheduling time, for example pod affinity, which shouldn't be checked when considering resize, because the pod is already scheduled. It's similar to our discussion about the new extension point which will be used by the workload-aware preemption to check whether the pods still fits on the node.

Reviewer 5: In the first place though, which plugins should be involved in the deferred pod preemption and how would the scheduler know it should use only those plugins for deferred pods? Like, I guess it is only the resource fit plugin within in-tree plugins but otoh we should not hard-code it, like "executing the resource fit plugin only when this is a deferred pod" would be bad.
Maybe we can just let pod-affinity/topology-spread plugins return Skip at PreFilter for deferred pods before calculating those things? i.e., in general, all plugins would be able to decide whether they will/will not influence deferred pod evaluation or not.

Author: I looked more closely today and I think I am more aligned with @sanposhiho's suggestions. The original PoC essentially would have hardcoded special-casing for deferred resize pods in many spots within the framework to ensure certain plugins are skipped or hardcoding the list of plugins to run or skip.

I'm working on an updated PoC that does something more like this:

> Maybe we can just let pod-affinity/topology-spread plugins return Skip at PreFilter for deferred pods before calculating those things? i.e., in general, all plugins would be able to decide whether they will/will not influence deferred pod evaluation or not.

I think I mostly agree with this take; plugins can each choose what they support / don't support. But I think it is cleaner to use the `ResizeInterestedPlugin` interface to allow plugins to declare whether they support resizing pods or not (rather than to modify every other plugin to return 'Skip' ). So essentially my proposal is to keep the `ResizeInterestedPlugin` interface as follows:

```go
// ResizeInterestedPlugin is an optional interface that plugins can implement 
// to declare interest in participating in the scheduling cycle for a deferred resize.
type ResizeInterestedPlugin interface {
	// ShouldHandleDeferredResize allows a plugin to opt-in or opt-out 
	// of the scheduling cycle for a deferred resize.
	ShouldHandleDeferredResize(ctx context.Context, pod *v1.Pod, nodeName string) bool
}
```

Then, the only plugins that implement this interface would be the node name plugin, the resource fit plugin, and the default preemption plugin. The framework would be modified to check whether a resize is deferred, and if so, skip all plugins that don't implement this interface.  The node name plugin additionally would be updated to implement the pre filter interface to restrict to just the pod’s nodeName.  I pushed a commit here for what this looks like: https://github.com/natasha41575/kubernetes/commit/0e41c0aa68e75436aa9583020645238db422f441

This has the following benefits:
- decouples the framework from hardcoding plugin names (i.e. DefaultPreemption, NodeResourceFit, NodeName)
- is more maintainable (most of the plugins are untouched)
- custom or out-of-tree scheduling plugins can seamlessly opt into deferred resize cycles by simply implementing this interface (and are opt-out by default)

Let me know what you think. I’m working on incorporating it into the PoC, but I think it is much cleaner than before. 

I updated the corresponding section of the kep in attempt to more clearly capture the details: https://github.com/natasha41575/enhancements/blob/dccdad9db779786768167b4a3c597a13c43638e9/keps/sig-scheduling/5836-scheduler-preemption-for-ippr/README.md#processing-deferred-resizes-in-the-scheduling-cycle

Reviewer 5: I am still thinking about this. I saw [CanPlaceBack](https://docs.google.com/document/d/1qMjV0Uwq5J9g2cYriWnLzDekaswc6dLA1h9GU4pCTtI/edit?pli=1&tab=t.0#heading=h.uqgpx0ellv1n) proposal on the other doc after leaving the last comment.
Like @macsko said, the motivation is _kinda_ similar. I am thinking if we can have a single new interface that satisfies what we need here and what WAP needs. (cc @Argh4k)

> But I think it is cleaner to use the ResizeInterestedPlugin interface to allow plugins to declare whether they support resizing pods or not (rather than to modify every other plugin to return 'Skip' ). 

PreFilter + Skip is a known established pattern for PreFilter plugins to declare "this pod is not relevant to me" by Skip status. So, I don't really feel a strong need to explicitly have a different interface just to say "I am not interested in deferred pod scheduling".
But, like I said, if there's a unified solution that satisfies us here and also replaces CanPlaceBack, that could be better than PreFilter + Skip though.

Reviewer 5: cc @wojtek-t too.

Author: > PreFilter + Skip is a known established pattern for PreFilter plugins to declare "this pod is not relevant to me" by Skip status. So, I don't really feel a strong need to explicitly have a different interface just to say "I am not interested in deferred pod scheduling".
> But, like I said, if there's a unified solution that satisfies us here and also replaces CanPlaceBack, that could be better than PreFilter + Skip though.

Ack, if PreFilter + Skip is the recommended pattern I'm happy to follow it. 

I took a look at the WAP proposal you linked (thanks, it was super helpful). There are a few reasons why I would hesitate to unify it right now:

- The design for canPlaceBack appears very specific to the reprieval process (PostFilter phase), but here we need modifications to all 3 phases (PreFilter, Filter, and PostFilter phases).
- The WAP proposal still seems to have some open questions about which plugins it needs to run, which may lead the use cases and design to evolve. I think it would help both features move forward if they are not coupled together while still in early phases. Particularly I think moving WAP to beta is a high priority, so I'd rather not force it to consider this preemption-for-ippr feature yet.
- The use case we have here is a bit simpler so I'd prefer not to tie it to something that is still being finalized.

For alpha I'm happy to take your suggestion to use PreFilter + Skip. I have it as a beta-blocker to reconsider the design in the context of workload-aware preemption. I think we can reassess whether and how to unify or modify the mechanism for WAP to suit this use case at that point in time.

Reviewer: > But, like I said, if there's a unified solution that satisfies us here and also replaces CanPlaceBack, that could be better than PreFilter + Skip though.

That's a great callout, but I'm afraid that even though these usecases look similar, they differ enough so that it will appear to be unnatural to merge them - let me explain why.
It's pretty clear to me that a set of plugins that both usecases will need will be different. NodePorts is a good example - it will definitely be needed to check in reprieval case, and we can clearly ignore it in IPPR preemption case. In theory we can hide the logic to distinguish these cases in a single function (like "if pod has spec.nodeName set return true; otherwise return false"), but that seems harder to reason about.

I acknowledge (and agree) that we should try to avoid creating new extension points if possible and try to limit their number, but merging it to a single extension point would make it very hard to explain which suggests it's not a good idea imho.
So I would suggest decoupling those and proceed with these two separately.

No getting to the usecase of IPPR - I think the crucial point about how we want to proceed is this part of Maciek's comment:

> Not really. PreFilter always has to be executed with all nodes, because it has to compute correct (anti-)affinity and topology spreading counts. Even if the NodeName plugin would return a single node in PreFilter, other plugins will execute as well, just the Filter will run on that one node.

And so the question is if we can mitigate the performance impact related to it.
So the two alternatives that we have most boil down to:

1) PreFilter + Skip - it keeps the plugin interface simpler, at the cost of:
- the need to extend most of plugins with the skipping logic in their PreFilter to avoid performance impact
- (out of core) plugins that don't reimplement it will continue to influence the preemption

2) Dedicated extension point - it extents the plugin interface, but:
- only plugins interested in resizing need to implement this method
- (out of core) plugins that don't implement the new method will not influence the preemption

So in my opinion, the decision about the pattern to choose should depend on the question: "what should be a default behavior for a plugin if it won't make any action?" If we want to take it into account in preemption by default, we should go with (1), otherwise we should go with (2).

My personal preference is towards (1) [potential performance hit, but not ignore the plugin], but I don't have enough data about what those plugins can do and can be convinced otherwise.

/cc @macsko @Argh4k

Reviewer 5: > That's a great callout, but I'm afraid that even though these usecases look similar, they differ enough so that it will appear to be unnatural to merge them - let me explain why.
It's pretty clear to me that a set of plugins that both usecases will need will be different. NodePorts is a good example - it will definitely be needed to check in reprieval case, and we can clearly ignore it in IPPR preemption case.

I agree they are different, but the point is whether they actually differ "enough" - You are saying they do, and I am still not sure, even after your comment. Your nodeport scenario is certainly one example for one plugin needing to behave differently between IPPR and WAP. But, that is all about whether we could nicely abstract a new interface so that it can accommodate two behaviors in a clear way, or that is difficult and would be a bad interface. This "there is a difference in necessary behaviors" is not really a right reason for not abstracting things because the abstraction is actually the weapon to wrap such difference.

Right now, CanPlaceBack is defined like this on [PoC](https://github.com/kubernetes/kubernetes/compare/master...Argh4k:kubernetes:inplace-reprieval)-

```go
CanPlaceBack(ctx context.Context, victimPod *v1.Pod, nodeInfo fwk.NodeInfo, clusterNodes []fwk.NodeInfo, preemptorPods []*v1.Pod) *fwk.Status
```

I totally understood the motivation with CanPlaceBack. But, as I mentioned on slack, I actually felt, and am still feeling, that this interface might be too specific to the current WAP usecase, like not as general as other classic interfaces that we have. The concern is that we might find it hard to expand later, or use it for other usecases later.

What WAP needs (this PoC's CanPlaceBack) is essentially "we placed these new pods (`preemptorPods`). Can we still run this pod (`victimPod`)?". And, what IPPR needs is "we made this running pod bigger. Can we still run this pod?". 
And, I actually see a future yet-another similar usecase, which is descheduling - "_something_ (any changes that descheduling should react to) happened somewhere, can we still run this pod?". 
We can see a certain common pattern with those three use cases. 

To give you a rough picture, one example interface to accommodate all use cases would be like- 

```go
// plugins will return the list of ClusterEventWithCanPlaceBack, similarly to ClusterEventWithHint (QHint).
type ClusterEventWithCanPlaceBack struct {
	// ClusterEvent could be AssignedPod/Add or AssignedPod/UpdatePodScaleUp only, for now.
	// Some plugins may want to do Pod/All to react to both events in the same manner, exactly like QHint, while other plugins may be interested in only one event.
	Event ClusterEvent

	// CanPlaceBackFn shows the pod can still run after this event happens.
	CanPlaceBackFn CanPlaceBackFn
}

// CanPlaceBackFn checks whether this pod can still run after the event.
type CanPlaceBackFn func(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) *fwk.Status
```

WAP can use CanPlaceBackFn with AssignedPod/Add, and IPPR preemption can use CanPlaceBackFn with AssignedPod/UpdatePodScaleUp. It might be useful for other more-complicated preemption scenarios that need to handle resources different from running pods.
And, descheduling will be able to use it similarly to QHint, like each plugin that wants to implement descheduling will react to each cluster event, and if CanPlaceBackFn returns Unschedulable, the scheduler deletes the pod to trigger the rescheduling.

... but again, I am not saying ClusterEventWithCanPlaceBack-ish unified interface is the best option. I am just thinking out laud and showing the option - just wanted to say that, _if we want_, we can create such a general interface to solve those different-but-similar use cases. As I've kept saying, I am still not sure what is the best in my mind.

Reviewer: > This "there is a difference in necessary behaviors" is not really a right reason for not abstracting things because the abstraction is actually the weapon to wrap such difference.

To clarify - I'm all for abstracting. If you have suggestions how to better abstract those things, I'm all ears - I absolutely agree that the current proposals aren't perfect.
What I was trying to say, that unless we find a common framing for these two things (which I personally can't see), we shouldn't try to abstract two separate things under one common abstraction.

> I totally understood the motivation with CanPlaceBack. But, as I mentioned on slack, I actually felt, and am still feeling, that this interface might be too specific to the current WAP usecase, like not as general as other classic interfaces that we have. The concern is that we might find it hard to expand later, or use it for other usecases later.

+1 - I just don't have a better counterproposal for it.

> To give you a rough picture, one example interface to accommodate all use cases would be like-

That's super interesting, because I was personally thinking about very similar thing, but decided that it's more complex than targeted interfaces. Not because the interface is complex, but because I believe that for a given plugin the implementations for different events (if needed) will be pretty disjoint - which suggests to me that this may not be the right abstraction.

But I share your primary concern of having gazillion dedicated targeted interfaces in the scheduling framework. I just didn't yet come up with anything that seemed natural to me.

Reviewer 4: I am pretty much on Wojtek's side here. I would like to have this as a neat abstraction, but I cannot come up with  one. As the input for discussion I think that describing canPlaceBack as "we placed these new pods (preemptorPods). Can we still run this pod (victimPod)" is not 100% accurate. The problem is that we are at the stage when we want to schedule preemptorPods. So the question is really, "we want to schedule new pods (preemptorPods) at this place. Can we still run this pod (victimPod) in its original place, so it does not break the schedulability promise?". It's more specific and the main difference is that we need to care also about the schedule constraints of preemptorPods that we no longer enforce when the pods are running. In the Beta update I proposed to name it just Reprieve and have it as another method on Filter plugins: https://github.com/kubernetes/enhancements/pull/6106

I do not think that we have an idea on how to share this interface with other actions.

Author: > I am pretty much on Wojtek's side here. I would like to have this as a neat abstraction, but I cannot come up with one. 

+1. I like the idea of unifying them but existing canPlaceBack seems very specific to WAP and I don't have a good idea how to unify them.

I understand though not wanting a dedicated extension point for every special case. I updated the KEP to state that we will use the Prefilter+Skip pattern on all plugins irrelevant to pod resize.

Reviewer 5: and now PreFilter+Skip makes the most sense because we no longer have CanPlaceBack for WAP

> binding evaluation, but will restrict its node evaluation search exclusively to the node the deferred pod is already running on. 

Reviewer 3: I think the scheduler shouldn't do binding for deferred pods? How would that look like otherwise?

Author: Yeah you are right; it should skip most of the plugins except for resource fit checks and preemption. Updated it accordingly.

In most cases, this restricted search will result in a `FitError`, which allows the Scheduler to naturally cascade into the preemption phase. However, if a `FitError` 
is not received (e.g., if cluster topology shifts free up space on the target node) and the resize now appears to fit, we will still move the deferred resize pod to the 
`Unschedulable` queue. This ensures that we maintain full tracking of the deferred pod resize request until the `Deferred` condition is completely removed by the 
Kubelet. The precise necessity for keeping the pod parked in the `Unschedulable` queue in these scenarios is described in more detail below in [Reconsideration Race Conditions](#reconsideration-race-conditions).

Within the preemption plugin, the search for victims must be restricted to the same node as the `Deferred` pod. Due to workload-aware preemption, though the space needs 

> to be freed up on the same node as the `Deferred` pod, this can result in preemption of pods on other nodes. 

Reviewer 5: > ue to workload-aware preemption, though the space needs 
to be freed up on the same node as the `Deferred` pod, this can result in preemption of pods on other nodes. 

Please elaborate on this. I do not understand

Reviewer 5: given the node name is set to deferred pods, I thought the preemption won't pick up different nodes

Author: IIUC, if the victim pod is part of a workload, workload aware preemption means that all the pods on that workload would be preempted.

The victim pod itself will always be selected from the same node.

Author: I added this based on @wojtek-t's feedback in https://github.com/kubernetes/enhancements/pull/5932#discussion_r2864053625.

Reviewer 5: ah ok thanks, now it is clear

Furthermore, the Scheduler needs to temporarily remove the `Deferred` pod from the node snapshot during calculations to ensure its current resource usage doesn't 
interfere with the preemption logic.

Taking all this into account, the logic for processing a `Deferred` resize is as follows:

1. **Identify Deferred Status**: Confirm the pod has a `Deferred` resize and is already bound to a node.

> 2. **Evaluate Fit and Trigger Preemption**: Perform normal node evaluation restricted to the current node. If the resize fits on the node, move the pod into the Unschedulable queue and skip victim selection (see [Reconsideration Race Conditions](#reconsideration-race-conditions) for more details).

Reviewer 2: I think it won't be normal node evaluation, as we don't have to apply pod-affinity (and similar) restrictions

Reviewer 4: +1, IIUC the whole effort from scheduler perspective is only about resources which are only one of the Filter plugins. If we check all plugins we can hit issues with scheduling constraints that might no longer hold true. For example if a pod was scheduled on this node because of pod affinity rule and the pod that made it possible was removed, pod affinity filter would fail, but this should not stop the pod from running (and upscaling) as this affinity is only a scheduling constraint.

Reviewer 2: So we could use the interface that @natasha41575 proposed (`ResizeInterestedPlugin` interface) to filter out any plugins that are not resize aware yet (mainly out-of-tree ones) or not relevant.

Eventually we should achieve a similar behavior that we need for reprieve.

Author: Completely agreed. I wrote it somewhere in the KEP that we should skip affinity / other constraints that are only applicable to scheduling. I'll change the wording here to make that more clear (instead of describing it as 'normal').

> So we could use the interface that @natasha41575 proposed (ResizeInterestedPlugin interface) to filter out any plugins that are not resize aware yet (mainly out-of-tree ones) or not relevant.

Using the proposed interface to filter out all irrelevant plugins SGTM. Thanks for the feedback!

Author: Updated the wording to specify that we only want to do the resource fit checks and not others like affinity etc

3. **Trigger Preemption**  If a `FitError` occurs, initiate the Scheduler preemption logic.
4. **Isolate Node**: Within the Preemption plugin, narrow the victim search exclusively to the pod's node, noting that this may still result in pods on other nodes being preempted due to workload-aware preemption.

> 5. **Snapshot Adjustment**: Temporarily remove the `Deferred` pod from the node snapshot to calculate required space accurately.

Reviewer 5: when and how? If we simply remove, the scheduler could misunderstood and could schedule some other new pods there.

Author: Ack, I updated this to avoid temporary removal. I explained more in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3269654047 and in the kep how to avoid double counting by just modify the resource fit plugin to not additionally include the resizing pod's resources to the node snapshot again. Hopefully the explanation makes sense; marking this as resolved for now.

> 6. **Calculate Victims**: Identify suitable preemption victims and then restore the pod to the snapshot. The preemption logic should run only the logic for the resource-fit check, skipping filters that are relevant only to initial scheduling, such as affinity/anti-affinity rules and topology spread constraints.

Reviewer 3: Could we reuse the extension point that will be added to the workload-aware preemption that will allow us to check whether the pod is still fitting the node? We wouldn't need to care about filtering the affinities and other plugins then. @dom4ha @Argh4k

Author: waiting for input from @dom4ha and @Argh4k because I'm not too familiar with the workload-aware preemption proposal. 

But I think it might be unnecessary if we proceed with the discussion in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3269191956. We still need to modify the node name and preemption plugins to be aware of how to handle resizes, so additionally making other plugins (affinity etc) opt-out of handling deferred resizes wouldn't add very much extra complication (depending on how we do it).

Author: Per the discussion in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3335584230, I think consensus is that using a unified extension point between this KEP and WAP is not desirable. I'm going to mark this conversation as resolved, but we can continue discussing in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3335584230 for further points.

7. **Update Status**: Report the success or failure of the preemption attempt in the pod status using the new ResizeUnschedulable condition. The pod should then be added to the Unschedulable queue to keep it tracked until the `Deferred` condition is removed by the Kubelet.

> 8. **Reevaluation**: When the victim pod is removed, the scheduler will be triggered to move all Unschedulable pods (including the Deferred resize pod) into the scheduling queue, resulting in reevaluation.

Reviewer 3: How will the state machine look like when the scheduling cycle says the pod is schedulable and the preemption is not required?

Author: This is written above:

> 2. **Evaluate Fit and Trigger Preemption**: Perform normal node evaluation restricted to the current node. If the resize fits on the node, move the pod into the Unschedulable queue and skip victim selection (see [Reconsideration Race Conditions](#reconsideration-race-conditions) for more details).

The `Reconsideration Race Conditions` section is intended to explain why moving it into the Unschedulable queue is necessary (as opposed to just removing pod from the queue) - let me know if something about it is not clear.

Author: These two sections are intended to answer your question:

> In most cases, this restricted search will result in a `FitError`, which allows the Scheduler to naturally cascade into the preemption phase. However, if a `FitError` 
is not received (e.g., if cluster topology shifts free up space on the target node) and the resize now appears to fit, we will still move the deferred resize pod to the 
`Unschedulable` queue. This ensures that we maintain full tracking of the deferred pod resize request until the `Deferred` condition is completely removed by the 
Kubelet. The precise necessity for keeping the pod parked in the `Unschedulable` queue in these scenarios is described in more detail below in [Reconsideration Race Conditions](#reconsideration-race-conditions).

and

> 2. **Evaluate Fit and Trigger Preemption**: Perform node evaluation restricted to the current node. The node evaluation logic should run only the logic for the resource-fit check, skipping filters that are relevant only to initial scheduling, such as affinity/anti-affinity rules and topology spread constraints. If the resize fits on the node, move the pod into the Unschedulable queue and skip victim selection (see [Reconsideration Race Conditions](#reconsideration-race-conditions) for more details).

The `Reconsideration Race Conditions` section is intended to explain why moving it into the Unschedulable queue is necessary (as opposed to just removing pod from the queue) - let me know if something about it is not clear.

### Scheduler Restart and State Recovery

Because the scheduler's internal `activeQ` and `Unschedulable` pods pool are maintained
in-memory, a scheduler restart clears the state of all pods currently undergoing preemption
for a resize. To prevent these pods from remaining in a `Deferred` state indefinitely,
the scheduler must proactively re-identify them upon startup.

*   **Cold-Start Re-Queueing:** Upon initialization and completion of the initial cache sync,

> the scheduler will perform a one-time scan of all pods with the `Deferred` condition.

Reviewer 3: Wouldn't it just be fixed by adding the deferred pods to the scheduling queue in eventhandler.AddPod method?

Author: Yeah sorry this is phrased poorly, I'll update it

*   **Re-evaluation:** Once in the queue, these pods will undergo the standard evaluation flow
(Node Fit -> Preemption).

*See risk considerations for shifting victims [here](#shifting-preemption-victims-during-scheduler-restart) and alternative considerations for persistent state [here](#persistent-preemption-state-across-restarts).*

### Scheduler Resource Reservation

Because the Scheduler uses max(spec, allocated, actual) when determining resource fit, Kubelet allocation of existing 
resize requests will take precedence over re-scheduling of the evicted pod. This means that the Scheduler does not

> need to take any special action to ensure the resources are reserved.

Reviewer 5: I don't understand at all about this part.

Author: see discussion in https://github.com/kubernetes/enhancements/pull/5932#discussion_r2864168606

Reviewer 5: Ok, so it is not talking only about "re-scheduling of the evicted pod", but all the scheduling attempts that will happen after the preemption. Can you rephrase the explanation here a little bit?

Author: rephrased - if it is still unclear please let me know

### Kubelet-Scheduler Preemption Interaction

No changes are required to the Kubelet. Currently, the Kubelet monitors pod removals, including evictions, and will 
automatically retry a resize as soon as it detects that a victim pod has been cleared.

> #### Reconsideration Race Conditions

Reviewer 3: Can the same pod be resized while its previous resize is deferred? If yes, could we have a section covering this scenario?

Author: The answer is yes, but it is related to your above comment: https://github.com/kubernetes/enhancements/pull/5932#discussion_r3219339318. This is the scenario I was referring to when I said that UpdatePod should watch for resource changes.

What we would want to do:
1. If the pod is sitting in the Unschedulable queue, move to the active queue
2. If the pod in the active queue already, update the pod in the active queue
3. If the pod is in the middle of the scheduling cycle when this happens, complete the scheduling cycle with the current pod snapshot. Then add it back to the active queue. I think this is just another case of (1) though, because after the scheduling cycle, the pod will be moved to the Unschedulable queue anyways.

Based on your comment above though:

> It may be tricky to implement to move the pod to the activeQ from within the event handler based on a specific event.

The proposal is from within `UpdatePod` we check if the resources have changed; then we can just remove the pod from the Unschedulable queue (if it's there), and then add it to the active queue. It looks like there is already an existing function ([schedulingQueue.Activate](https://github.com/kubernetes/kubernetes/blob/d8dc1e25b01e08916652fb89f600319f95571937/pkg/scheduler/backend/queue/scheduling_queue.go#L101-L106)) that implements this. I pushed a commit to demonstrate what this looks like in code: https://github.com/kubernetes/kubernetes/commit/47b549db42a614b5df98260efbf66c08faa049b7

Author: I updated this section of the kep to more clearly cover this scenario: https://github.com/natasha41575/enhancements/blob/dccdad9db779786768167b4a3c597a13c43638e9/keps/sig-scheduling/5836-scheduler-preemption-for-ippr/README.md#changes-to-updatepod-eventhandler

details about the failure.

The Scheduler will remove this condition when the pod is no longer in the `Deferred` state, or if it has a successful preemption attempt in the future.

#### Handling retries

To ensure efficiency and avoid unnecessary reconciliation loops, the Scheduler will treat `Deferred` resizes that fail preemption as a 
standard `FitError` (Unschedulable). By classifying the resize failure this way, the existing Scheduler retry and backoff 
logic is triggered automatically, requiring no new dedicated retry infrastructure.

1.  **Queue Placement**: When a preemption attempt fails to find enough victims on the node to satisfy the resize, the pod is moved to 
the `unschedulablePods` pool within the scheduling queue.
2.  **Triggering Re-entry**: Because the pod is in the `unschedulablePods` pool, it will be moved back to the `activeQ` (or 
`backoffQ`) automatically when cluster events occur that could potentially change the outcome of the preemption logic. These events 
include:
    * **Pod Deletion**: A pod on the same node is deleted or removed (freeing up capacity).
    * **Pod Downsize**: A pod on the same node has its resource requests decreased.

>     * **Node Update**: The node's `Allocatable` resources increase (e.g., via a Kubelet config update).

Reviewer 5: For this, do we need to support Qhint for PostFilter plugins? that is not mentioned on KEP.

Reviewer 2: +1 The KEP mentions that but does not calls that QHints needs to be update explicitly

Reviewer 3: Why do we need to support them? The pod will be rejected by the Filter, populating the `UnschedulablePods` set. We don't need any QHints for preemption itself

Reviewer 5: It depends on https://github.com/kubernetes/enhancements/pull/5932#discussion_r3194157047 and we should come back to this point after the thread concludes.

Author: Yes, I think we should use QHints precisely to ensure only events on the same node as the resizing pod triggers it to be reevaluated. I updated the KEP to state this explicitly.

### Preventing breaking custom PostFilterPlugins

Currently, scheduler preemption is implemented as a PostFilterPlugin. While the [PostFilterPlugin interface](https://github.com/kubernetes/kubernetes/blob/ab1a54911670bbc1792062c5ed5fbcc3df9dfcb1/staging/src/k8s.io/kube-scheduler/framework/interface.go#L534) 
allows for custom implementations, existing customs plugins may not handle `Deferred` resizes correctly. To prevent 
breaking changes, plugins must be able to communicate to the Scheduler whether they support Deferred resize logic,
with the default being `false`.

We propose a new, optional interface that PostFilterPlugins can implement to signal this support:

```go
// ResizeInterestedPlugin is an optional interface for PostFilter plugins.
type ResizeInterestedPlugin interface {
    // ShouldHandleDeferredResize allows a plugin to opt-in or opt-out 
    // of the scheduling cycle for a deferred resize.
    ShouldHandleDeferredResize(ctx context.Context, pod *v1.Pod, nodeName string) bool
}

> ```

Reviewer 5: I don't understand why we need this. Can't we change the default binding plugin just skip making a bind API call when it is a deferred pod?

Reviewer 2: IIUC some plugins should not be executed as we don't want to re-evaluate not scheduling-time constrains like pod-affinities and run into PostFilter because of them.

So only the plugins that knows how tho deal with Pod-resizes would be executed.

Reviewer 3: > Can't we change the default binding plugin just skip making a bind API call when it is a deferred pod?

I think we should stop the happyy path scenario just after scoring (or even filtering) without executing further steps like reserve, permit, prebind etc.

Author: I explained a little bit more in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3269191956 and also tried to clarify in the KEP. The goal is to avoid having to either modify every plugin (besides the relevant ones) to be a no-op in case of deferred pods, or to have to hardcode in the framework what plugins to skip.

Author: Took this out, so the conversation is no longer relevant. We can continue discussing in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3335584230.

Therefore, the decision to have the Kubelet manage this policy via the Pod status (rather than 
the Scheduler reading the Node annotation directly) is driven by two factors: first, maintaining 
a pod-centric scheduler preemption design; and second, ensuring the policy remains strictly 
coupled to the resize lifecycle (which Kubelet owns) rather than general scheduling.

Instead, the mechanism relies on cooperation between the Kubelet and the Scheduler via the Pod status:
1.  **Configuration**: A cluster operator (or an autoscaler) adds the annotation `scheduler.policy/disable-pod-resize-preemption: "true"` to a Node.
2.  **Kubelet Inspection**: The Kubelet watches its own Node object and caches this configuration.
3.  **Pod Status Update**: When the Kubelet evaluates an In-Place Pod Resize request and determines it must be `Deferred` due to insufficient capacity, it checks the node annotation.
4.  **Signaling the Scheduler**: If the annotation is set to `"true"`, the Kubelet updates the Pod's status by setting a new Kubelet-owned condition: `ResizePreemptionDisabled` to `True` (with Reason: `NodePolicyEnforced` and an appropriate message).

> 5.  **Scheduler Action**: The Scheduler, in its `UpdatePod` event handler and scheduling queue processing, will inspect the Pod's conditions. If a `Deferred` pod has `ResizePreemptionDisabled: True`, the Scheduler **will not** add the pod to the scheduling queue for preemption retry, and will ignore it for preemption purposes. The pod will remain `Deferred` without triggering cluster disruption.

Reviewer 5: Why the scheduler has to check `ResizePreemptionDisabled: True` as well? Kubelet should not set `Deferred` if the annotation is set, in the first place?

Author: The pod is still considered `Deferred` from Kubelet perspective. Kubelet needs to set `Deferred` regardless of whether the annotation is set. The annotation is just to provide a means to tell the **scheduler** not to preempt. Other controllers still need to know that the pod is in Deferred state (for example, VPA). We cannot change this; it would be a breaking change to a GA feature.

Btw, this was discussed in the sig scheduling meeting and this is the guidance I got around it.

ETA: I reread the KEP and realized the discussion about kubelet behavior may have sounded ambigious. I'll update it to explicitly say this.

Reviewer 5: > The pod is still considered Deferred from Kubelet perspective. Kubelet needs to set Deferred regardless of whether the annotation is set.

I think this part confused me - other than the scheduler, is there anyone else who reads this "Deferred" condition and needs to take some action?

Author: Yeah; for example there are external autoscalers such as VPA, Ray Autoscaler or the GKE Cluster Autoscaler that are relying on the `Deferred` condition. They may want to take decisions to trigger an autoscaling event regardless of scheduler preemption behavior.

Even if we had some opinion about how the autoscalers should interact with scheduler preemption (which I stated as a non-goal), keeping the `Deferred` resize in the status imo would still be both low-cost, and necessary, as it is already GAd as a part of the pod API and we should be very careful not to change the meaning.

Reviewer 5: I see. But, I am not sure if VPA future interaction should be on non-goal completely. Because such future enhancement could depend on a design decision we will make here. We should be, at least, confident that what we do here is aligned with such future enhancement.

I am foreseeing a potential future that we want to trigger VPA only when the scheduler cannot preempt any other running pods for deferred pods. That is aligned to the current behavior that CA started to provision a node only when a pod has once attempted to be scheduled and failed.
Then, VPA would have to look at ResizeUnschedulable condition so that it only targets deferred pods that cannot be scaled up by the preemption. (This is also aligned to the current behavior that CA looks at Unschedulable condition to target pods that have failed at scheduling once) 

And, coming back to the topic here - I am wondering if we can just use ResizeUnschedulable condition with a different reason when the resize preemption is disabled, instead of a dedicated ResizePreemptionDisabled condition. Because if we use `ResizePreemptionDisabled: True` condition like you are proposing (and if my foreseeing future actually comes true), VPA would have to look at pods that have _either_ ResizeUnschedulable or ResizePreemptionDisabled because the scheduler would completely ignore pods with ResizePreemptionDisabled and would not set ResizeUnschedulable. It looks cleaner to unify the same condition is used when a deferred pod cannot be scaled up by the preemption, regardless of whether it is because there's no pod to preempt or because the preemption is disabled in the first place.

Author: > I see. But, I am not sure if VPA future interaction should be on non-goal completely. Because such future enhancement could depend on a design decision we will make here. We should be, at least, confident that what we do here is aligned with such future enhancement.
> 
> I am foreseeing a potential future that we want to trigger VPA only when the scheduler cannot preempt any other running pods for deferred pods. That is aligned to the current behavior that CA started to provision a node only when a pod has once attempted to be scheduled and failed.
> Then, VPA would have to look at ResizeUnschedulable condition so that it only targets deferred pods that cannot be scaled up by the preemption. (This is also aligned to the current behavior that CA looks at Unschedulable condition to target pods that have failed at scheduling once)

Yeah, that makes sense; I think we are aligned. The design should not preclude what we think are reasonable future interactions, and I have the same vision of a potential interaction with autoscalers - I actually have an internal doc that makes a very similar proposal :) 

For the time being, I'd like kubelet to continue to set the `Deferred` condition regardless, allowing time for the autoscalers to make that migration, but it sounds like overall we are directionally aligned. 

> It looks cleaner to unify the same condition is used when a deferred pod cannot be scaled up by the preemption, regardless of whether it is because there's no pod to preempt or because the preemption is disabled in the first place.

I like the suggestion to reuse `ResizeUnschedulable`. I agree that it is cleaner, and semantically it is equivalent from the perspective of external controllers. I was a little worried about a condition having two different components (kubelet and scheduler) writing to it, but as I think a little more, this shouldn't be an issue if we designate kubelet as the sole owner when pod resize preemption is disabled and scheduler as the sole owner when pod resize preemption is enabled. I'll think on this a little more, but I'll update the KEP accordingly.

Author: I updated the KEP to propose reusing `ResizeUnschedulable`, with a note that scheduler will own it when preemption is enabled and kubelet will own it when it is disabled. If it looks reasonable to you, I'll mark this discussion as resolved. Thanks for the suggestion!

Author: After API review with @thockin, I think this discussion is no longer relevant as I took out the scheduler-owned part of the condition. Further context is in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3335559305.

*   **Why Rejected**: This violates the core scheduler design principle that preemption logic should be pod-centric, not node-centric. It couples the scheduler to specific node configurations and complicates the preemption plugin. Furthermore, since this policy applies *only* to resize and not initial scheduling, embedding this check directly in the scheduler's generic preemption path is inappropriate. It also prevents the Kubelet from honoring the policy for its own internal preemption.

#### 2. Node Labels (Rejected)
*   **Description**: Setting the policy via a node label, e.g., `kubectl label node <node-name> scheduler.policy/disable-pod-resize-preemption="true"`.
*   **Why Rejected**: Labels are intended for identifying and grouping objects (e.g., for node selectors or affinity). They are not meant for carrying fine-grained configuration or policy intent. Annotations are the standard Kubernetes mechanism for attaching non-identifying metadata and configuration, making them a better fit.

#### 3. Node Field (Rejected)
*   **Description**: Introducing a new field within the `Node` spec or status, e.g., `spec.preemptionPolicies.disablePreemptionForPodResize: true`.

> *   **Why Rejected**: This requires a core Kubernetes API change to the `Node` object. Such changes are heavyweight, require extensive review, and increase API churn. Since this policy is specific to the In-Place Pod Resize feature and primarily benefits specific environments (resizable nodes), introducing a permanent field in the core `Node` API is not justified when annotations provide sufficient extensibility.

Reviewer 5: I don't understand. annotation is not an API. I feel like you are just choosing the annotation because you feel troublesome to go thru the API change review and decided to use the annotation for the sake of "bypassing" the review.

I really don't like a core component depending on an annotation, generally speaking

Reviewer 5: If this is "this policy is specific to the In-Place Pod Resize feature and primarily benefits specific environments", then you should reconsider if we need to put this feature to k/k in the first place. I see kinda contradiction here - you are saying this benefits only specific envs and cannot put to API, while you are trying to put to k/k.

Author: > you feel troublesome to go thru the API change review 

I don't know why you are making this assumption that I am suggesting this out of laziness. I suggested it because it's what seems best based on my understanding. I even took this to a SIG Scheduling meeting to get feedback about the API, but no one voiced any concerns about this there. I'm open to being wrong and accepting feedback about that though, especially if an API approver suggests one of the alternatives.

I'm perfectly happy to put it in an API field if that's what is recommended for this kind of behavior. I actually don't have a preference between the different options, so I'm happy to follow upstream guidance.

> for the sake of "bypassing" the review

I have been working in upstream k8s long enough to be aware that this KEP needs to go through an API review regardless of whether it's an annotation or not... I have been planning to ask one this week to take a look.

Author: Discussed offline with @tallclair - I will rework the KEP to change the proposal from an annotation to a full API on the node.

Author: I updated the proposal to be a full API on the node. PTAL when you get a chance.

  * We would have to define the matrix of interactions between two independent preemption fields, leading to ambiguous states 
  (e.g. what happens if `preemptionPolicy` is set to `Never` but `resizePreemptionPolicy` is set to `PreemptLowerPriority`)?
  * Adds code complexity, including updates to the Pod API, validation logic, and new branches in the Scheduler.

If a pod is marked with `preemptionPolicy: PreemptLowerPriority`, the user has already communicated that this workload
is more important than lower-priority tasks. Whether that importance is manifested during initial placement or vertical 
scale-up, the intent of the priority remains the same.

*See alternative considerations for preemption policies [here](#preemption-policies-1).*

### Pod Priority, Graceful Termination, and Pod Disruption Budget

Because we are reusing the existing Scheduling preemption logic, the behavior will be aligned with existing preemption
behavior for new pods. This means:

  * Pods will be preempted according to their Pod Priority.
  * The graceful termination period of victim pods will be honored.
  * PodDisruptionBudget is supported, but not guaranteed.   

> ### Node-level Preemption Policy for In-Place Pod Resize

Author: I had an offline API review with @tallclair and @thockin. Summarizing the feedback from discussion here (for posterity):

- Call the top-level struct `PodPreemptionPolicy` as we may want to have a different API for preemption policies for other objects
- Add 'label-format' validation to the strings under the preemption policy and update the examples to have a k8s.io prefix
- Don't have the scheduler set a `PodResizeUnschedulable` condition (yet). We don't yet have a concrete use case in mind for how such a condition would be used; for now the scheduler can surface the results of the scheduling cycle via events for observability. We can revisit setting such a condition in the future as we further design the interactions between higher-level controllers and this resize preemption behavior.

I updated the KEP to take the above feedback; it's ready for another pass.

To avoid unnecessary disruption on nodes that have their own autoscaling solutions (such as resizable nodes), cluster operators and controllers need a way to disable scheduler-triggered preemption specifically when it is caused by an in-place pod resize request on those nodes. This protects critical workloads by preventing the scheduler from preempting low-priority pods to make room for a higher-priority resize, allowing the system to prefer node upsizing over preemption-induced disruption.

We propose a new field on the `Node` object to configure this policy, wrapped within a high-level `PreemptionPolicy` struct in `NodeSpec`:

```go
// NodeSpec contains the desired configuration of a node.
type NodeSpec struct {
    // PodPreemptionPolicy controls the node-level preemption behaviors for pods on this node.
    // +optional
    PodPreemptionPolicy *NodePreemptionPolicy `json:"podPreemptionPolicy,omitempty"`
}

// NodePreemptionPolicy defines the node-level policies governing preemption for pods on this node.

> type NodePreemptionPolicy struct {

Reviewer 6: Should probably be NodePodPreemptionPolicy

Author: done!

    // DisableResizePreemption lists the owners (e.g., autoscalers, operators, administrators) 
    // that have requested to disable scheduler and Kubelet preemption for in-place pod resize on this node.

>     // If this list is non-empty, resize-induced preemption is disabled on this node.

Reviewer 6: Spec the validation:

format
how many items can be here

@lalitc375 @yongruilin -- are we open to DV on Node yet?

Reviewer 7: DV has been wired and enforced to all types. If they are new fields, they can directly use DV with tags and  without adding any handwritten validation. 
Proposed DV tags in a separate comment.

Author: thanks! added the DV tags

> type NodePreemptionPolicy struct {
>     // DisableResizePreemption lists the owners (e.g., autoscalers, operators, administrators) 
>     // that have requested to disable scheduler and Kubelet preemption for in-place pod resize on this node.
>     // If this list is non-empty, resize-induced preemption is disabled on this node.
>     // +listType=set
>     // +optional
>     DisableResizePreemption []string `json:"disableResizePreemption,omitempty"`

Reviewer 7: ```suggestion
    // +listType=set
    // +optional
    // +k8s:optional
    // +k8s:eachVal=+k8s:format=k8s-label-key
    // +k8s:maxItems=<int> 
    // +k8s:listType=set
    DisableResizePreemption []string `json:"disableResizePreemption,omitempty"`
```

Author: added!

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### How Deferred Resizes Integrate into the Scheduling Queue

The scheduling queue is shared between new pods and `Deferred` pods, using existing queuing logic to sort them by their current scheduling priority. 

*See alternative considerations for separate scheduling queues [here](#separate-scheduling-queue) and prioritized resize logic [here](#implementing-prioritized-resizes-logic).*

#### Detecting Deferred Resizes in UpdatePod

The [UpdatePod event handler](https://github.com/kubernetes/kubernetes/blob/60433d43cf0bb83a2ac7d5e767137b3d510026ec/pkg/scheduler/eventhandlers.go#L147) is responsible for maintaining the state of the scheduling queue based on changes to the pod objects, and includes logic to handle pods with a `Deferred` resize:

* **If the deferred condition is added to a pod**: The Scheduler adds the pod to the scheduling queue.

> * **If the pod's spec (desired) resources change**: The Scheduler updates its view of the pod, and moves it to the active queue if it is not already there. This can happen if the `Deferred` pod's desired resources are updated before the previously requested resize can complete.

Reviewer 3: What about the backoffQ? Should we always move the pod to the activeQ in this case or honor the backoff time?

Reviewer 3: Will such move be done by the QHint of NodeResourcesFit plugin or differently? Preferably, it should be a QHint. If this behavior requires adjusting the NodeResourcesFit's QHint, please state that in the KEP

Author: > What about the backoffQ? Should we always move the pod to the activeQ in this case or honor the backoff time?

IMO it should be moved to the activeQ in this case. The pod's resources changing is a new user-initiated resize request, so I think that should be treated as resetting the backoff.

> Will such move be done by the QHint of NodeResourcesFit plugin or differently? Preferably, it should be a QHint. If this behavior requires adjusting the NodeResourcesFit's QHint, please state that in the KEP

Sorry, maybe I'm misunderstanding your suggestion - why does this specifically need to be done by the QHint? I was envisioning having the `updatePod` event handler making a call to schedulingqueue.Activate as demonstrated in this commit: https://github.com/kubernetes/kubernetes/commit/47b549db42a614b5df98260efbf66c08faa049b7.

Reviewer 3: > Sorry, maybe I'm misunderstanding your suggestion - why does this specifically need to be done by the QHint? I was envisioning having the `updatePod` event handler making a call to schedulingqueue.Activate as demonstrated in this commit: [kubernetes/kubernetes@47b549d](https://github.com/kubernetes/kubernetes/commit/47b549db42a614b5df98260efbf66c08faa049b7).

We rather try to avoid adding more use cases of external activations as it complicates the code and is prone to non-obvious issues. In particular, if something is achievable using QHints, it should use them. It looks like using NodeResourcesFit QHint for this should work. If not, we can consider activating the pod from the event handler.

Author: understood, thanks for the explanation. I updated this section to specify that this change would be detected by QHints.

*See alternative considerations for how to handle this without dual representation (such as using a dynamic `PodGetter` interface) [here](#conflicting-sources-of-truth-for-the-pod-1).*

### Processing Deferred Resizes in the Scheduling Cycle

When processing a pod with a `Deferred` resize, the scheduler runs it through a modified scheduling cycle. Rather than evaluating the pod as if it were brand-new to the cluster, the cycle is restricted exclusively to resource capacity checks and preemption logic. This differs from standard scheduling in two key ways:

* **Targeted Node Search**: Candidate node evaluation is restricted solely to the node the deferred pod is already running on.
* **Bypassing Irrelevant Constraints**: All scheduling plugins irrelevant to direct resource capacity fitting (such as node affinity, pod anti-affinity, and topology spread constraints) are bypassed.

To achieve this cleanly and without hardcoding specific plugin names on the framework side, we adjust all irrelevant plugins to be skipped when processing `Deferred` pods.

> Only **`NodeName`**, **`NodeResourcesFit`**,  and **`DefaultPreemption`** plugins implement handling for processing `Deferred` pods. All other plugins are modified to implement the `PreFilter` interface and return `Skip` for `Deferred` pods, which results in the plugin being bypassed for deferred resize scheduling cycles.

Reviewer 3: Add that out-of-tree plugins developers will have to be advised to align their implementation for `Deferred` pods.

Author: added!

#### `NodeName` Plugin: Node Restriction via the PreFilter Phase

> The `NodeName` plugin is adjusted to implement the `PreFilter` phase, and is the only plugin that runs during the `PreFilter` phase for deferred resize pods.

Reviewer 3: I think that `NodeResourcesFit` should also run its` PreFilter` for deferred pods. I know that it can do the relevant precomputation in the `Filter`, but doing so would only complicate the code

Author: Ack, updated the KEP accordingly that it will run its `PreFilter` and `Filter` for deferred pods.

Because the deferred resize pod is already bound to and running on a host, the scheduler limits the candidate search space at the start of the scheduling cycle. If a pod specifies `Spec.NodeName`, its `PreFilter` returns a `PreFilterResult` containing only that node. This restricts the candidate node evaluation search space to just the pod's currently assigned node.

#### `NodeResourcesFit` Plugin: Calculating Resource Fit in the Filter Phase

The `NodeResourcesFit` plugin is the only plugin that runs during the `Filter` phase for deferred resize pods. It runs a modified resource-fit check to specially handle `Deferred` pods.

##### Ignoring `desired` resources for other pods

**For Scheduling New Pods**

The Scheduler (via the `NodeResourcesFit` plugin) assumes the resources of pods to be `max(desired, allocated, actual)`. This ensures that the node does not end up overcommitted in the event that Kubelet actuates a resize while the new pod is being scheduled. This behavior is correct for scheduling of newly created pods.

**For Resize of Existing Pods**

> The Kubelet determines resource fit in the event of a resize by assuming resources as: 

Reviewer 2: I think that computing resources differently depending on the context would unnecessarily complicate the scheduler logic, as it could not reuse the existing mechanism of taking max(desired, actual).

It'd be also against what this KEP says in "Race between a Deferred resize and a new higher-priority pod": 

> From the scheduler's view, once the spec is updated, the resources are already reserved. It doesn't matter from the scheduler perspective whether the resize has been actuated by the Kubelet yet.

Note that we will most likely drop the idea of introducing Reprieve extension point in WAP (https://github.com/kubernetes/enhancements/pull/6106#discussion_r3372371793), which means that the reprieve process may rerun filter on victims. It means that victims assessment (in Filter) would not know the context (whether it's a Filter for preemptor that use max or Filter for victim that should not use it).

Reviewer 3: > which means that the reprieve process may rerun filter on victims

In our recent idea, in WAP the reprieval should run the Filter only on preemptors. So it shouldn't be a problem

Author: I stated this in https://github.com/kubernetes/enhancements/pull/5932#discussion_r3381981158 as well, but copying it here because it is relevant to the discussion. 

For newly scheduled pods, we need to continue using max(desired, allocated, actual). It is by design (not a bug). This is because the kubelet is constantly checking if desired != allocated, can it allocate the desired resources. So it may at any point, independent of the scheduler, update the allocated resources.

This results in a race condition where, if the scheduler ignores desired resources, the scheduler may try to bind the pod to the node at the same time the kubelet allocates a new resize. If kubelet-resize wins the race, kubelet will then reject the newly scheduled pod during kubelet pod admission.

> It'd be also against what this KEP says in "Race between a Deferred resize and a new higher-priority pod"

This part is specifically discussing the race between scheduling a _new_ pod. 

Scheduler-evaluation of already-existing `Deferred` resize pods do not suffer from this race condition because they are already bound to the node. There is no new pod being added and possibly overcommitting a node, because all pods in question are already bound and running.

So we are left just with discussing implementation complexity. Initially, I was thinking:
1. We already have a special-case ignoring `desired` resources; this is done for infeasible resizes: https://github.com/kubernetes/kubernetes/blob/e1d80088da7453801a8a8d1de3e81720982e2ac6/staging/src/k8s.io/component-helpers/resource/helpers.go#L300-L301. So there is precedence for this kind of behavior.
2. Kubelet reuses the scheduling libraries when determining resource fit. It solves the problem by looping through other pods on the node and overwriting the 'spec' resources to the 'allocated' resources: https://github.com/kubernetes/kubernetes/blob/e1d80088da7453801a8a8d1de3e81720982e2ac6/pkg/kubelet/allocation/allocation_manager.go#L618-L627. 

I took a closer look at the scheduling framework though and I can see now that it is a more complicated change than it seems. I hadn't realized that the `max(desired, allocated, actuation)` calculation happens as a precomputational step to populate `NodeInfo`, and that conditionally changing whether to consider `desired` or not will require a much larger change. 

So I am aligned with @dom4ha's suggestion to leave this out of scope here. But I want to add it to the known risks, and consider it as something we may want to fix in the future. I will update the KEP accordingly - thank you for the catch!

ETA: I do want to spend some time next cycle exploring if there is a way to handle this without adding significant complexity, but I don't think we need to block on it.

Reviewer 3: I agree to leave this out of scope for alpha and reconsider whether we really need this during beta discussion

Reviewer 2: Agree as well. I guess we could have the behavior you described but as you noticed, the cost of having it may simply not pay off.

IMO the most reasonable way to address the problem long-term is to let scheduler decide which pod should be resized and communicate it somehow to kubelet. It should be some mechanism equivalent to NNN that would be later honored and confirmed by kubelet. This way we'd reduce the kubelet freedom to manage resources only to non-reserved resources, but changing the already confirmed reservation (NNN) would have to go though the scheduler again (change NNN if desired).

We need that mechanism anyway to make sure the scheduler decisions for larger workloads (where the cost of scheduling and binding is high) won't be altered. So in the meantime, we should live with taking `max(desired, allocated, actuation)` on deferred pods with all its consequences.

- For the pod that is being resized, use `max(desired, allocated, actual)`. 

> - For all other pods on the node, use `max(allocated, actual)` (ignoring desired). 

Reviewer 3: `max(desired, allocated, actual)` is the actual behavior in all cases, right? IIUC, we can have the issue described below even now, when some (non-deferred) high priority pod wants to preempt deferred pods. So, it sounds even like a bugfix for the actual behavior

Author: > So, it sounds even like a bugfix for the actual behavior

For newly scheduled pods, we need to continue using max(desired, allocated, actual). It is by design (not a bug fix). This is because the kubelet is constantly checking if desired != allocated, can it allocate the desired resources. So it may at any point, independent of the scheduler, update the allocated resources.

This results in a race condition where, if the scheduler ignores desired resources, the scheduler may try to bind the pod to the node at the same time the kubelet allocates a new resize. If kubelet wins the race, it will reject the pod during kubelet pod admission.

`Deferred` resize pods do not suffer from this race condition because they are already bound to the node.

Reviewer 3: Right, it fully makes sense, thanks for clarifying

Author: But per https://github.com/kubernetes/enhancements/pull/5932#discussion_r3382162797 I think this is a moot point, as I will be taking it out.

If we use our modified logic, where only the resizing pod uses `max(desired, allocated, actual)` and the remaining pods use `max(allocated, actual)`, then:

1. Preemption logic first removes all candidate pods, so it sees node is using X (Pod 1's max).
2. The Scheduler first tries to reprieve pod 2, which assumed to have max(X/2, X/2) resources.  X/2 + X = 1.5X, which is less than node allocatable 2X, so pod 2 is reprieved.
3. The Scheduler tries to reprieve pod 3, which assumes to have max(X/2, X/2) resources. X/2 + 1.5X = 2X, which still fits within the node allocatable 2X, so pod 3 is reprieved.
4. The Scheduler tries to reprieve pod 4, which assumes to have max(X/2, X/2) resources. X/2 + 2X = 2.5X, which is more than node allocatable 2X, so pod 4 is not reprieved.

This results in only 1 pod (Pod 4) being preempted, which is the correct behavior.

##### Avoiding double counting

To prevent the double-counting of resources when a deferred-resize pod is evaluated for scheduling or preemption fit on its assigned node, the `NodeResourcesFit` plugin is adjusted to leverage the scheduler cache's existing resource accounting logic. 

> If the pod is already assigned to the node being evaluated, the scheduler's cache has already factored the pod's resource footprint into the node's aggregated requested resources (`nodeInfo.Requested`). Rather than adding the pod's new `desired` resources to the `nodeInfo.Requested`, which would count the pod twice, the plugin instead uses the maximum of the new `desired` resources and the existing `nodeInfo.Requested` for the pod.

Reviewer 3: It can happen that during scheduling of the deferred pod, the pod taken from the queue will be different than pod taken from the cache (because of the timing). How will such divergence be handled?

Author: > It can happen that during scheduling of the deferred pod, the pod taken from the queue will be different than pod taken from the cache (because of the timing). 

Can you help me understand when this can happen? Pod updates should always be propagated to both the pod taken from the queue and the pod taken from the cache.

But in any case, if there is a divergence, I think we can handle it by always taking the max. I can add a note to the KEP on this if you are able to help me better understand the scenario you have in mind.

Reviewer 3: > > It can happen that during scheduling of the deferred pod, the pod taken from the queue will be different than pod taken from the cache (because of the timing).
> 
> Can you help me understand when this can happen? Pod updates should always be propagated to both the pod taken from the queue and the pod taken from the cache.

When the scheduling cycle starts for the (deferred) pod, the pod object stays the same throughout the whole cycle. When the pod update arrives to the scheduler and pod is already popped to the scheduling cycle, it won't be updated, but its cached representation will be. It would lead to pod object inconsistency, which has to be consistently handled.

> But in any case, if there is a divergence, I think we can handle it by always taking the max.

Sounds like a good option. I think we would also need to ensure the pod is updated first in the cache and then in the queue to only have the potential pod object divergence in one way (so the cache representation cannot be older than queue/scheduling cycle representation).

Author: Sounds good. I added both points to the kep (order of updates and how to handle the divergence). Thank you!

##### Handling Node Evaluation Results

In most cases, this restricted search results in a resource fit error, naturally triggering the scheduler's preemption pathway. In this case, the pod will be kept in the `Unschedulable` queue.

> However, if the resize fits without needing preemption (e.g., if cluster topology shifts or other workloads exit, freeing up capacity on the host), the scheduler skips the preemption phase, but keeps the deferred pod in the `Unschedulable` queue.

Reviewer 3: To put the pod into the `Unschedulable` queue it should have the plugin that rejected it. In this case, it wouldn't have any plugin. Without the plugin set, no QHints will work for the pod, meaning it will stay in the queue until flush happens. Shouldn't the pod be moved to the backoffQ instead?

Reviewer 3: This also means that we will need a branch in scheduling cycle in which we'll abort continuing the processing when the filters are successful. You can mention in which place the execution will stop. Directly after the Filter?

Author: > To put the pod into the Unschedulable queue it should have the plugin that rejected it. In this case, it wouldn't have any plugin. 

I updated the KEP to clarify that the NodeResourcesFit plugin should reject it and mark it as `UnschedulableAndUnresolvable` with an appropriate message. This will cause the pod to be moved to the Unschedulable queue while still bypassing the preemption plugin.

> Shouldn't the pod be moved to the backoffQ instead?

I think Unschedulable queue is correct over backoffQ. There is no need to reevaluate the pod periodically. If kubelet is able to actuate the resize, the pod will lose its `Deferred` condition and scheduler can remove it from the Unschedulable queue. If kubelet is not able to actuate the resize, the victim removal will cause the Deferred pod to be moved to the activeQ. 

> This also means that we will need a branch in scheduling cycle in which we'll abort continuing the processing when the filters are successful. You can mention in which place the execution will stop. Directly after the Filter?

I think this question no longer applies if we are considering this case as a NodeResourceFit plugin failure.

Reviewer 3: Idea of using `UnschedulableAndUnresolvable` looks good. We also have a `Pending` status that I think could be used here, but it looks like it would still trigger the preemption

Author: > We also have a Pending status that I think could be used here, but it looks like it would still trigger the preemption.

Another option is to use Pending, and have the DefaultPreemption plugin have a special case to skip if the status is Pending && the pod is deferred. But this sounds unnecessarily complicated to me.

The necessity of keeping the pod in the `Unschedulable` queue in both cases is explained further in [Reconsideration Race Conditions](#reconsideration-race-conditions).

#### `DefaultPreemption` Plugin: Preemption Mechanism Adjustments in the PostFilter Phase

> Only the `DefaultPreemption` plugin runs during the PostFilter phase for deferred resize pods. No additional adjustments are needed, as the modifications to the `NodeName` and `NodeResourceFit` plugins already ensure the desired behavior:

Reviewer 3: How will that be handled that only the `DefaultPreemption` plugin will be run at PostFilter? Will other PostFilter plugins self-disable themselves similarly to Filter?

Author: Yeah, added a note to state this explicitly.

* **Victim Selection Constraint**: The search for preemption victims is isolated exclusively to the deferred pod's active node. This is already done by the `NodeName` plugin in the `PreFilter` phase. However, because of workload-aware group scheduling, preempting a victim on the target node might still trigger the cascaded preemption of related pods on other nodes.

> * **Resource Accounting Adjustment**: During the preemption evaluation, the fit plugin is adjusted to specially handle the deferred pod's resources, as described in [Calculating Resource Fit in the Filter Phase](#calculating-resource-fit-in-the-filter-phase).

Reviewer 3: Link to `Calculating Resource Fit in the Filter Phase` is outdated. ditto in two places below

Author: my bad! thank you for catching this; updated it.

2. **Evaluate Fit and Trigger Preemption**: Perform node evaluation restricted to the current node. 
    * The node evaluation logic should run only the logic for the resource-fit check, skipping filters that are relevant only to initial scheduling, such as affinity/anti-affinity rules and topology spread constraints. 
    * The resource-fit logic is adjusted to specially handle the deferred pod's resources, as described in [Calculating Resource Fit in the Filter Phase](#calculating-resource-fit-in-the-filter-phase).
    * If the resize fits on the node, move the pod into the `Unschedulable` queue and skip victim selection (see [Reconsideration Race Conditions](#reconsideration-race-conditions) for more details).
3. **Trigger Preemption**  If a `FitError` occurs, initiate the Scheduler preemption logic.
4. **Calculate Victims**: Identify suitable preemption victims, with the search restricted to the current node. The resource-fit logic is adjusted to specially handle the deferred pod's resources, as described in [Calculating Resource Fit in the Filter Phase](#calculating-resource-fit-in-the-filter-phase).

> 5. **Reevaluation**: When the victim pod is removed, the scheduler is triggered to move all `Unschedulable` pods (including the `Deferred` resize pod) into the scheduling queue, resulting in reevaluation.

Reviewer 2: ```suggestion
5. **Reevaluation**: When the victim pod is removed, the `Deferred` resize pod is triggered and moved from `Unschedulable` pods into the scheduling queue, resulting in reevaluation.
```

Not all pods are triggered, just the ones assessed to be potentially schedulable.

Author: updated!

> 5. **Reevaluation**: When the victim pod is removed, the scheduler is triggered to move all `Unschedulable` pods (including the `Deferred` resize pod) into the scheduling queue, resulting in reevaluation.

Reviewer 3: I don't see anywhere a note what steps will be skipped from the scheduling cycle failure handler. In particular, do we want to set any pod status? If yes, which one? AFAIK, we also don't want to set the NNN of the deferred pod, right?

Author: Re the pod status: I had a discussion with the API approver and the decision was that we will not expose any pod status in the case of failure. We will just emit events for observability.

But thanks yeah I missed specifying what will be skipped in the failure handler. I have added a section under `#### Failure Handler Adjustments for Deferred Pods`

Reviewer 3: Sounds good. New section looks good as well

### Kubelet-Scheduler Preemption Interaction

The Kubelet monitors pod removals, including evictions, and automatically retries a resize as soon as it detects that a victim pod has been cleared. The removal of the victim pod triggers a scheduling cycle, during which the deferred resize pod is reevaluated.

#### Reconsideration Race Conditions

It is possible that the Kubelet chooses to resize a different pod than the one that triggered the eviction. Consider
the case where Pod A has a deferred resize and triggers the eviction of Pod B. In the meantime, a higher-priority resize of 
Pod C is requested. In this case, the kubelet correctly prioritizes the resize of Pod C over the resize of Pod A.
In this case, we ensure that Pod A stays in the scheduling queue until its resize can complete successfully, 
including any additional evictions it may trigger.

We can model deferred resize retries analogous to the way standard scheduling already works today. 
For standard pods, if a pod cannot be scheduled:
1. The new pod triggers preemption.
2. The pod gets moved to the `Unschedulable` queue.

> 3. When the preemption victim is finally removed, all unschedulable pods get moved back to the active (or backoff) queue.

Reviewer 2: to be precise, not all unschedulable pods move back, but the ones that may become schedulable

Author: updated, thanks for the catch!

When the Node's preemption policy prohibits preemption for resize requests (the `spec.podPreemptionPolicy.disableResizePreemption` list is non-empty), the Kubelet sets the `PodResizePreemptionDisabled` condition to `True` with reason `PreemptionDisabledByNodePolicy` upon deferring the resize. This communicates to the scheduler that the `Deferred` resize request is blocked because preemption has been disabled on the node.

```yaml
status:
  conditions:
  - type: PodResizePreemptionDisabled
    status: "True"
    reason: PreemptionDisabledByNodePolicy
    message: "Preemption for in-place pod resize is disabled on node 'node-1' by preemption policy."
    lastTransitionTime: "2026-02-23T15:23:13Z"
```

The Kubelet clears this condition when the node's preemption policy is updated to allow preemption (clearing the list), or when the pod's `Deferred` resize is resolved or cancelled.

##### Scheduler Action

The Scheduler ignores `Deferred` pods that have the `PodResizePreemptionDisabled` condition set to `True` with reason `PreemptionDisabledByNodePolicy`. 

The Scheduler's `UpdatePod` handler watches for pod updates and reacts to the `PodResizePreemptionDisabled` condition as follows:

> *   **Condition Added**: When the `PodResizePreemptionDisabled` condition is set with reason `PreemptionDisabledByNodePolicy`, the Scheduler removes the pod from the scheduling queue.

Reviewer 3: Does the scheduler has to check for exact reason? Isn't `PodResizePreemptionDisabled=True` condition enough?

Author: > Isn't PodResizePreemptionDisabled=True condition enough?

Yeah, should be sufficient. Updated accordingly. Thanks!

Reviewer 3: I think it's not yet updated in the line "Condition Added"

Author: fixed!

This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
For more details, see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/testing-strategy.md

If integration tests are not necessary or useful, explain why.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, document that tests have been written,
have been executed regularly, and have been stable.
This can be done with:
- permalinks to the GitHub source code
- links to the periodic job (typically https://testgrid.k8s.io/sig-release-master-blocking#integration-master), filtered by the test name
- a search in the Kubernetes bug triage tool (https://storage.googleapis.com/k8s-triage/index.html)
-->

We will create integration test(s) to ensure basic functionalities:

- A lower-priority pod can be preempted by a higher-priority `Deferred` resize.
- Several lower-priority pods can be preempted by a higher-priority `Deferred` resize.

> - Non-preempting pods according to the `preemptionPolicy` do not trigger preemption.

Reviewer 3: It would be good to increase the integration test coverage for beta, as the feature is extensive. For alpha, the listed tests should be enough

Author: SGTM, thanks! I added a note here that we need to expand the integration test coverage and will finalize the full integration test list prior to beta.

and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

<!--
Describe them, providing:
  - API call type (e.g. PATCH pods)
  - estimated throughput
  - originating component(s) (e.g. Kubelet, Feature-X-controller)
Focusing mostly on:
  - components listing and/or watching resources they didn't before
  - API calls that may be triggered by changes of some Kubernetes resources
    (e.g. update of object X triggers new updates of object Y)
  - periodic API calls to reconcile state (e.g. periodic fetching state,
    heartbeats, leader election, etc.)
-->

> For most of the logic, we are not introducing any new API calls. The Scheduler is already watching for pod updates. The node and pod states are cached in the Scheduler's memory, and is updated when the Informer receives an update. The kubelet will update the pod status when the resize is `Deferred` due to insufficient capacity; this is not a new API call. 

Reviewer 3: If kube-scheduler would set any status saying the deferred pod triggered preemption, then the new API calls will be added. I'm not sure, though, whether we need such status or not.

Author: Per the discussion with the API approver, we will not be adding a new status. I think when we go to beta, we should define some events to surface the results of the scheduler decision for the sake of observability, but we don't plan to expose it in pod status.

Reviewer 3: Ack

<!--
Describe them, providing:
  - API type
  - Supported number of objects per cluster
  - Supported number of objects per namespace (for namespace-scoped objects)
-->

No.

###### Will enabling / using this feature result in any new calls to the cloud provider?

<!--
Describe them, providing:
  - Which API(s):
  - Estimated increase:
-->

No.

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

<!--
Describe them, providing:
  - API type(s):
  - Estimated increase in size: (e.g., new annotation of size 32B)
  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
-->

The new condition in the pod status will increase the size of pod objects, only when the condition is present.

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

<!--
Look at the [existing SLIs/SLOs].

Think about adding additional work or introducing new steps in between
(e.g. need to do X to start a container), etc. Please describe the details.

[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
-->

> No.

Reviewer 3: It may delay scheduling of the pods, because deferred pods will be handled by the main scheduler loop

Author: ah you are right, let me add that here.

All 3 pods (Pod 2, 3, and 4) would be preempted, even though it was only necessary to preempt one of them.

To prevent unnecessary preemption, the `NodeResourcesFit` plugin would need to likewise assume resources in the same way as the Kubelet when evaluating deferred pods, while maintaining its existing behavior for scheduling new pods.

If we use our modified logic, where only the resizing pod uses `max(desired, allocated, actual)` and the remaining pods use `max(allocated, actual)`, then:

1. Preemption logic first removes all candidate pods, so it sees node is using X (Pod 1's max).
2. The Scheduler first tries to reprieve pod 2, which assumed to have max(X/2, X/2) resources.  X/2 + X = 1.5X, which is less than node allocatable 2X, so pod 2 is reprieved.
3. The Scheduler tries to reprieve pod 3, which assumes to have max(X/2, X/2) resources. X/2 + 1.5X = 2X, which still fits within the node allocatable 2X, so pod 3 is reprieved.
4. The Scheduler tries to reprieve pod 4, which assumes to have max(X/2, X/2) resources. X/2 + 2X = 2.5X, which is more than node allocatable 2X, so pod 4 is not reprieved.

This results in only 1 pod (Pod 4) being preempted.

> Such a modification to the `NodeResourceFit` plugin is out of scope for this KEP due to its wide ranging implications across the scheduler. However, it may be considered as an effort separate from this KEP in the future.

Reviewer 8: I am not sure if below case is covered. 

Consider a node with allocatable memory = 2X.

Pod1 is high priority:
- desired = X
- allocated = X/2
- actual = X/2
- resize state = Deferred

Pods 2/3/4 are lower priority:
- desired = X/4
- allocated = X/2
- actual = X/2
- resize in progress

(This is a general case for a controller to resize some pods using less resources for other pods on the same node to increase their resources.)

Using the rule above, it may decide that Pod1’s deferred resize requires preemption of one pod from 2/3/4.

Concern: this design may trigger premature or unnecessary preemption when other pods on the node are already in the process of shrinking and are expected to release resources soon.

Author: This is true, but I think this is not specific to this KEP. The scheduler even during regular scheduling may choose to preempt a pod on a node that is in the process of shrinking other pods, meaning that it is expected to release resources soon. So I would suggest to take this as a separate issue (do you have time to file one and CC me?)

Reviewer 8: > The scheduler even during regular scheduling may choose to preempt a pod on a node that is in the process of shrinking other pods, meaning that it is expected to release resources soon. 

I think it makes sense to treat this separately. To make the boundary clearer in the KEP, I can add an explicit note that this KEP does not try to make preemption account for resources that may be released soon by other pods whose shrink resize is already in progress, since that is a broader scheduler limitation rather than something specific to deferred-resize preemption. Not sure if this should be included in the `Limitation` or `Non Goals` section. 

What I was trying to call out is that this KEP also inherits that limitation when deferred-resize preemption is evaluated in the presence of other pods that are already shrinking via in-progress in-place resize.

That said, I’m okay leaving it out of scope for this KEP and tracking it separately if reviewers prefer. Or let’s wait a bit for additional feedback; otherwise I can file a follow-up issue and link it here then.

Author: > (This is a general case for a controller to resize some pods using less resources for other pods on the same node to increase their resources.)

One of the reasons I suggest to file a separate issue is because I want to hear more about this controller use case and I worry we will lose track of it if we leave the discussion here. In general I would advise such a controller that wants to do this kind of modification to first downsize other pods prior to requesting the upsize. Many things can go wrong during pod resize and it is not guaranteed to succeed. 

But if this is not possible and the controller instead wants to react to `Deferred` resize, this is exactly the use case we had in mind in the "Node Level Preemption Policy" section - the policy is introduced specifically to support custom autoscaling controllers. The idea we have is that autoscaling controllers can disable ippr-preemption while they are doing their own autoscaling adjustments, and can dyamically re-enable ippr-preemption if they need help from the scheduler to make room for the resize. I am curious to hear if this policy is sufficient for the use case you have in mind as well. 

If not, we may consider some other solution. One idea is if there are downsizes in progress that would make enough room for the pod, skip preemption. But I think that is a broader discussion that would apply to both scheduling and resize.

> Such a modification to the `NodeResourceFit` plugin is out of scope for this KEP due to its wide ranging implications across the scheduler. However, it may be considered as an effort separate from this KEP in the future.

Reviewer: While I agree we shouldn't solve it for Alpha, shouldn't we actually attack this problem for Beta?
The unnecessary preemptions is something that users generally catch pretty rapidly and complain about very quickly.

I would suggest adding a Beta criteria to revisit this decision if this shouldn't be addressed.

Author: I do plan to look at this a little more next cycle, but I am not too sure that we will end up addressing this based on https://github.com/kubernetes/enhancements/pull/5932#discussion_r3392315276.

But SG to add it to at least discuss in Beta if this should be addressed. So I added it to the beta criteria.

#### `NodeName` Plugin: Node Restriction via the PreFilter Phase

The `NodeName` plugin is adjusted to implement the `PreFilter` phase. Because the deferred resize pod is already bound to and running on a host, the scheduler limits the candidate search space at the start of the scheduling cycle. If a pod specifies `Spec.NodeName`, its `PreFilter` returns a `PreFilterResult` containing only that node. This restricts the candidate node evaluation search space to just the pod's currently assigned node.

#### `NodeResourcesFit` Plugin: Calculating Resource Fit in the Filter Phase

The `NodeResourcesFit` plugin will run its `PreFilter` and `Filter` phases for deferred resize pods. It runs a modified resource-fit check to specially handle `Deferred` pods.

To prevent the double-counting of resources when a deferred-resize pod is evaluated for scheduling or preemption fit on its assigned node, the `NodeResourcesFit` plugin is adjusted to leverage the scheduler cache's existing resource accounting logic. 

> If the pod is already assigned to the node being evaluated, the scheduler's cache has already factored the pod's resource footprint into the node's aggregated requested resources (`nodeInfo.Requested`). Rather than adding the pod's new `desired` resources to the `nodeInfo.Requested`, which would count the pod twice, the plugin instead uses the maximum of the new `desired` resources and the existing `nodeInfo.Requested` for the pod.

Reviewer 9: >  existing `nodeInfo.Requested` for the pod 

This is a little confusing. `nodeInfo.Requested` is a single aggregate value. Do you mean the cached state of the pod from NodeInfo ?

Author: Yes, you are right. `nodeInfo.Requested` is a single aggregate value, and the current phrasing is confusing.

In the normal case, the scheduler's cache has already factored the pod's resource footprint into `nodeInfo.Requested` as `max(desired, allocated, actual)`, so the plugin will simply use `nodeInfo.Requested` directly.

However, if the pod object from the queue has resource requests that diverge from the pod object in the cache (due to race conditions or sync lag), the plugin will resolve this by finding the pod in `NodeInfo`, comparing the two footprints, and adjusting the node's aggregate requests by adding the difference (ensuring we use the maximum of the two). I rewrote this section to make it clear.

#### Summary of Scheduling Cycle Flow

Taking all the above into account, the logic for processing a `Deferred` resize is as follows:

1. **Identify Deferred Status**: Confirm the pod has a `Deferred` resize and is already bound to a node. Enqueue it in the Scheduling queue.
2. **Evaluate Fit and Trigger Preemption**: Perform node evaluation restricted to the current node. 
    * The node evaluation logic should run only the logic for the resource-fit check, skipping filters that are relevant only to initial scheduling, such as affinity/anti-affinity rules and topology spread constraints. 
    * The resource-fit logic is adjusted to specially handle the deferred pod's resources, as described in [`NodeResourcesFit` Plugin: Calculating Resource Fit in the Filter Phase](#noderesourcesfit-plugin-calculating-resource-fit-in-the-filter-phase).
    * If the resize fits on the node, move the pod into the `Unschedulable` queue and skip victim selection (see [Reconsideration Race Conditions](#reconsideration-race-conditions) for more details).
3. **Trigger Preemption**  If a `FitError` occurs, initiate the Scheduler preemption logic.

> 4. **Calculate Victims**: Identify suitable preemption victims, with the search restricted to the current node. The resource-fit logic is adjusted to specially handle the deferred pod's resources, as described in [`NodeResourcesFit` Plugin: Calculating Resource Fit in the Filter Phase](#noderesourcesfit-plugin-calculating-resource-fit-in-the-filter-phase).

Reviewer 2: We need to remember that during preemption and reprieve we run Filter only for ResourceFit the same as during Filtering phase, otherwise we'd take pod-affinity and other initial scheduling constraints which could block the preemption.

Author: thank you - added a note here on that.

For standard pods, if a pod cannot be scheduled:
1. The new pod triggers preemption.
2. The pod gets moved to the `Unschedulable` queue.
3. When the preemption victim is finally removed, pods that were previously unschedulable but may now become schedulable are moved back to the active (or backoff) queue.

With deferred resize, we can follow a parallel path:
1. The resize request triggers preemption.
2. The pod gets moved to the `Unschedulable` queue.
3. When the preemption victim is removed, the deferred resize pod is moved back to the active (or backoff) queue.

However, a notable race condition arises: the Kubelet may not finish resizing either Pod A or Pod C before the scheduler evaluates Pod A again. During this reevaluation, the scheduler observes that the resize now 'fits' on the node during its node fit checks. 

To protect against this race, we park the resize back into the `Unschedulable` queue. The lifecycle then resolves via one of three eventualities:
- **Pod A is correctly resized**: It loses its deferred condition. The scheduler observes this through watches and discards it from tracking.

> - **Pod C is resized**: (Even though the scheduler preempted for Pod A). There is now no longer enough room for Pod A's resize, which safely remains in the `Unschedulable` queue.

Reviewer: But at this point it's possible that we can trigger another preemption to make space for pod A.
To achieve that, we would need to move pod A to the activeQ, no?

@macsko - for your thoughts

Author: Ah you are right, thanks for the catch. The resize of Pod C should trigger Pod A to return to the activeQ.

IIUC, this can be achieved via QHints. We can add a QHint for deferred resize pods that if another pod on the node is upsized, we should move them to activeQ.

This may occasionally trigger a redundant scheduling cycle for other pods that don't fit, but this should be relatively rare, as the check would be strictly node-scoped and pod upsize events don't happen with significant frequency on a single node.

I added it to the KEP. LMK if this resolves the concern.

Reviewer: Yes - that's what I had on my mind exactly.

### Tracking Preemption Nominations to Avoid Double Preemption

If a `Deferred` resize triggers preemption, there is a risk of "double preemption" (where the scheduler terminates a different set of victims for the same resize request) in two scenarios:
1. **Scheduler Restart**: The scheduler restarts mid-preemption, clearing its in-memory preemption state, and selects a new victim pod upon recovery.
2. **Long Grace Periods**: A victim has a long graceful termination period. While it is terminating, other node activity (such as pod deletion) triggers the scheduler to re-evaluate the `Deferred` pod. If the cluster state has changed or selection is non-deterministic, the preemption logic might select a new victim instead of recognizing the existing termination process.

> To mitigate this, the scheduler needs a way to track which pod is preempting what via resource nominations so it does not retry preemption while the initial victim is still terminating. We considered three potential options to address this, and will finalize the approach during the Beta phase:

Reviewer: nit: there are now 4 options not 3

Author: fixed

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

In-Place Pod Resize requests that result in a `Deferred` resize status will trigger scheduler preemption, keeping the 
behavior aligned with scheduler preemption for newly created pods. 

The Scheduler will watch for `Deferred` resizes. To make room for the resize, it will try to preempt (evict) lower-priority pods to make the resize possible. The Kubelet today watches for pod removals, and will react to the eviction by reattempting the previously `Deferred` resize. 

If there are not enough lower-priority pods on the node to make space for the resize, the Scheduler will need to expose this 
information in the pod status. This would allow external controllers to provision additional capacity, if desired. This will 

> be exposed as a new `PodResizeUnschedulable` pod condition owned by the Scheduler, as described in the [PodResizeUnschedulable Pod Condition](#PodResizeUnschedulable-pod-condition) section. 

Reviewer 10: Do we need the scheduler to communicate to the controllers with the disable preemption policy mechanism? I thought the idea of the policy is that we give controllers priority for handling deferred resizes? I think having both mechanisms adds a lot of unnecessary complexity.

Author: I'm happy to just recommend that controllers use the disable-preemption-policy mechanism. I took out this sentence:

> This would allow external controllers to provision additional capacity, if desired.

But I think the scheduler still needs to report success/failure of the preemption attempt in the pod status for visibility.

resource deficit from an indefinite Deferred status into the active scheduling queue, preempted pods will naturally trigger cluster autoscalers to provision new capacity. 

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

#### Performance impact

If there are a significant number of deferred pods, periodic processing of those pods can affect scheduling throughput. To mitigate this, deferred pods will only be reconsidered if something on the node has changed that could make preemption succeed.

#### Interaction with workload-aware preemption

With workload-aware preemption, the resize may end up triggering preemption of a potentially large workload. In these cases, the preempted workload would always be lower priority than the pod that is being resized, so we consider this working as intended.

> #### Race between a Deferred resize and a new higher-priority pod

Reviewer 10: There's also a race condition with a subsequent resize:

1. Pod is resized
2. Resize is deferred
3. Pod is resized again
4. Scheduler sees deferred resize, along with resized resources from (3).

I think we can probably just ignore this case. If the requests in (3) are bigger than the requests in (1), then it will also be deferred, so that is the desirable case. If they're smaller and the revised resize wouldn't be deferred, then maybe the scheduler would see that it fits, and skip preemption?

If we were really worried about this case we could probably do something with ObservedGeneration, but I'm not sure it's necessary.

Author: > the revised resize wouldn't be deferred, then maybe the scheduler would see that it fits, and skip preemption

yeah I think this is handled already by the proposal that if the scheduler sees that the desired pod fits, it skips preemption

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### How Deferred Resizes will integrate into the Scheduling Queue

The scheduling queue will be shared between new pods and `Deferred` pods, using existing queuing logic to sort them by their current scheduling priority. 

*See alternative considerations for separate scheduling queues [here](#separate-scheduling-queue) and prioritized resize logic [here](#implementing-prioritized-resizes-logic).*

#### Detecting Deferred Resizes in UpdatePod

In the [UpdatePod event handler](https://github.com/kubernetes/kubernetes/blob/60433d43cf0bb83a2ac7d5e767137b3d510026ec/pkg/scheduler/eventhandlers.go#L147) that the Scheduler already has, the Scheduler will check to see if a pod has recently been marked as having a `Deferred` resize. If so, the pod will be added to the scheduling queue.

The `UpdatePod` event handler will also watch for:

> * **If the pod's resources changed from last time**: We need to update our view of the pod, and move it to the active queue if it is not there already. This can happen if the `Deferred` pod's desired resources are updated before the previously requested resize can complete.

Reviewer 10: I'm confused by this.

> If the pod's resources changed ...

Which resources? Desired resources? 

Is this saying, if the scheduler missed the first update when the spec changed, and it sees the deferred status and the spec update at the same time?

Author: Sorry, I'll update the wording to clarify.

It should say that we need to watch if the pod's **desired** resources changed from last time. 

If the pod is **already Deferred** and **already in the scheduling queue**, the scheduler needs to watch whether the pod's desired resources have changed so that it can (a) update its snapshot of the pod and (b) move it from the idle queue to the active queue

The primary risk with this approach is ensuring that no existing or future scheduler logic implicitly assumes a pod has a single unique representation or reference. We must audit the scheduler codebase to ensure that:
1. Pod updates correctly target both locations in tandem.
2. No downstream scheduling or preemption logic relies on strict pointer equality of pod objects across the cache and queue boundaries.
3. Resource double-counting is avoided on the target node by excluding the pod's additional resources during resource fit calculations as described in the section below.

*See alternative considerations for how to handle this without dual representation (such as using a dynamic `PodGetter` interface) [here](#conflicting-sources-of-truth-for-the-pod-1).*

#### Avoiding 'Double Counting' for Resource Fit and Preemption Plugins

To prevent the double-counting of resources when a deferred-resize pod is evaluated for scheduling or preemption fit on its assigned node, the `NodeResourcesFit` plugin is adjusted to leverage the scheduler cache's existing resource accounting logic:

> 1. **Assigned Node Evaluation**: If the pod is already assigned to the node being evaluated, the scheduler's cache has already factored the pod's maximum resource footprint (`max(desired, actuated, allocated)`) into the node's aggregated requested resources (`nodeInfo.Requested`).

Reviewer 10: When evaluating how many resources will be freed from a preemption, it should probably just look at allocated resources, since that's what kubelet uses to determine fit. If it takes the max, and the preempted pod has a deferred resize, it might not free up a sufficient amount.

I imagine this gets pretty messy in the scheduling context though. If we decide a pod doesn't fit because it's evaluating the total based on the max, so the scheduler initiates preemption, but then finds that no preemption is needed because now it's looking at just allocated... Maybe scheduling should only consider the max resources for pods of equal or higher priority? Not sure if the scheduling framework can accommodate that though.

I'm not sure how to handle this.

Author: You're right, thanks for calling out this case. I think this a side effect of the in-place pod resize design and it becomes messy by nature.

For the rest of my comment, I'm using the following terminology: 
- resizing pod = the pod being processed by the scheduler
- victim pod = the pod(s) being preempted
- candidate = the pod(s) under consideration for preemption

Preemption works as follows:
1. remove all candidate (lower-priority pods) from the node
2. reprieve (add back) the candidates one at a time, keeping them only if they fit
3. all candidates that do not fit receive an eviction request

IIUC what you want is:
- during the filter phase (standard scheduling cycle resource fit check), the scheduler needs to use max(desired, allocated, actual) 
- during the preemption phase, when calculating how much space is freed by the victim, use only the allocated resources for the _victims_, but the calculation for assumed resources for all non-candidate pods and all reprieved pods must still be max(desired, allocated, actual) to ensure the node isn't overcommitted. 

If we don't explicitly handle this case, what will happen is that when the victim pods are removed, the scheduler will reevaluate the resizing pod and see that it needs to preempt again.

I think we can handle this by modifying the reprieve process:
1. At the beginning, after all candidates are initially removed, the scheduler's 'assumed' resources for the node, should be sum((max(desired, allocated, actual)) of non-candidates + max(desired, allocated, actual) of the resizing pod
2. During reprieve process, we need to have two checks:
  a. The current reprieve fit check, where the fit check evaluates if the candidate can be added back using its max(desired, allocated, actual) footprint to ensure we don't overcommit the node.
  b. An additional check, that determines whether the sum of the remaining candidates' footprints based on _just their allocated_ resources frees up sufficient space for the resize. i.e. check that the `sum(remaining candidates' allocated) > resizing pod's desired - resizing pod's allocated`
3. If it passes 2a AND 2b, the pod is safely reprieved. If it fails either, the pod must be evicted. 

wdyt?

ETA: I do think we can / should solve this problem in this kep, but it occurred to me that we probably already have this problem with scheduler preemption for newly created pods today.

Author: Hmm... thinking more and I'm trying to reason about whether this is actually a problem. 

I am struggling to find an execution path where Check 2b changes the final preemption outcome. 

The only scenario I can think of where using max(desired, allocated) instead of allocated for the victim pod would result in not freeing up enough resources is when desired > allocated.

And if desired > allocated, I can't think of a scenario where Check 2a (the existing fit check) could pass but Check 2b would fail... Because the scheduler cache tracks the node's requested resources using max(desired, allocated, actual), the baseline snapshot is already overcommitted on paper whenever a deferred resize is present. It doesn't calculate how many resources are 'freed'; it calculates the max current requests against the node capacity.

It might help if you can help me come up with a scenario (with numbers) that illustrates when the issue could occur? I think I still want to add the extra check for defensive purposes but I think it is redundant

Reviewer 10: Before I was thinking about this as doing a relative diff: the scheduler decides in needs to free up `X` resources, so it preempts a pod with `{desired=X, allocated=X/2}`, but from the Kubelet's perspective this only frees up `X/2`.

In actuality, I think we have the opposite problem. Consider this case with 4 pods, all deferred:
- Pod1 (high priority): `desired=X, allocated=X/2`
- Pod2,3,4 (low priority): `desired=2X, allocated=X/2`
- Node `allocatable=2X`

The scheduler sees that the node is currently using `7X` based on the max resources. It preempts all 3 low priority pods (based on the max resources), when it only needed to preempt 1 (based on allocated resources).

Author: > In actuality, I think we have the opposite problem.

Ah that makes more sense; using max everywhere will conservatively tend to undercommit the node. 

In your scenario:

1. Preemption logic will first remove all candidate pods, so it sees node is using X (Pod 1's max). 
2. Scheduler tries to 'reprieve' the lower priority pods one at a time. First, it tries to reprieve pod 2 which is assumed to have max(2X, X/2) resources. 2X + X = 3X, which is more than node allocatable 2X. It will do the same for pod 3 and pod 4, so you are right it will preempt all 3 pods unnecessarily.

I think this means that the reprieve process should be modified so that candidate pods use allocated rather than max(desired, allocated).

I'm trying to think about whether non-candidate pods should also use allocated rather than max(desired, allocated) during the preemption phase. I think the answer is yes. Kubelet uses max(desired, allocated) only for the pod being admitted/resized and it uses allocated for everything else, so it would make sense for the preemption logic to align with that.

The scheduling framework can accommodate this with some tweaks; I will take a look and update the KEP accordingly.

Author: Updated the kep accordingly (with your example) that for resize, we should use allocated (ignoring desired) as the resource footprint of other pods on the node

#### Reconsideration Race Conditions

It is possible that the Kubelet will choose to resize a different pod than the one that triggered the eviction. Consider
the case where Pod A has a deferred resize and triggers the eviction of Pod B. In the meantime, a higher-priority resize of 
Pod C may be requested. In this case, the kubelet will correctly prioritize the resize of Pod C over the resize of Pod A.
In this case, we will need to ensure that Pod A stays in the scheduling queue until its resize can complete successfully, 
including any additional evictions it may trigger.

We can model deferred resize retries analogous to the way standard scheduling already works today. 
For standard pods, if a pod cannot be scheduled:
1. It triggers preemption.
2. It sets the standard `Unschedulable` condition.
3. It gets moved to the `Unschedulable` queue.
4. When the preemption victim is finally removed, all unschedulable pods get moved back to the active (or backoff) queue.

With deferred resize, we can follow a parallel path:
1. The resize request triggers preemption.

> 2. It sets a `PodResizeUnschedulable` condition.

Reviewer 10: What is `It` on this line? I think this is referring to the scheduler, but doesn't the Kubelet own the PodResizeUnschedulable condition?

Author: updated this to be more clear

Furthermore, this policy is strictly limited to pod resize requests and does not apply to new pod scheduling. In a mixed cluster (containing both resizable and non-resizable nodes), applying a preemption-disabling policy to new pod scheduling on a resizable node would cause the scheduler to simply select victims on a non-resizable node instead, defeating the purpose of the policy. For in-place pod resize, however, preemption is already strictly confined to the target node where the pod is running.

Instead, the mechanism relies on cooperation between the Kubelet and the Scheduler via the Pod status, leveraging the `PodResizeUnschedulable` condition which is described in more detail in the [PodResizeUnschedulable Pod Condition](#PodResizeUnschedulable-pod-condition) section.

The cooperation between the Kubelet and Scheduler is as follows:
1.  **Configuration**: A cluster operator or controller adds its identifier to the `spec.preemptionPolicy.disablePodResizePreemption` set on the Node object.

> 2.  **Kubelet Inspection**: The Kubelet watches its own Node object and caches this configuration.

Reviewer 10: It should also update any existing deferred resizes, right?

Author: yup! updated

We propose that this Kubelet-side preemption for resize **must also honor** the node-level preemption policy.
If the Node's `spec.preemptionPolicy.disablePodResizePreemption` list is non-empty, the Kubelet **must not** preempt existing pods on the node to accommodate the resize request of a critical pod. Instead, the Kubelet will mark the critical pod's resize as `Deferred` and set the `PodResizeUnschedulable` condition to `True` with reason `PreemptionDisabled`, just as it would for a non-critical pod.

This ensures a consistent operational model for the node: *no pods will be preempted to satisfy any resize request*, preserving the stability of all workloads on the node and forcing the system to rely on node autoscaling (upsizing) to resolve the resource deficit.

### Failures and Reconsideration of Deferred pods

#### PodResizeUnschedulable Pod Condition

To provide a unified view for external controllers, we propose a single new `PodResizeUnschedulable` condition type in the Pod status. This condition indicates that a `Deferred` resize request cannot proceed due to insufficient capacity and cannot benefit from scheduler-side preemption.

> The condition is co-owned by the Scheduler and the Kubelet under mutually exclusive, deterministic ownership regimes based on whether preemption is enabled:

Reviewer 10: I'm skeptical of this co-ownership working. Consider this case:

1. Scheduler evaluates preemption, cannot preempt, adds PodResizeUnschedulable (PRU) condition
2. An external controller determines it can handle resizes, and adds itself to the node's disablePodResizePreemption policy

Who owns the PRU condition now? The scheduler set it, but the node has now taken over ownership?

Reviewer 10: I see this is addressed below, but I'm still not sure it works well.

Author: I updated these sections to attempt to address some of your concerns around the co-ownership, and also explicitly walk through the cases of the preemption policy changes. 

It might be easier to discuss in our 1:1 later about whether this works well now.

```yaml
status:
  conditions:
  - type: PodResizeUnschedulable
    status: "False"
    reason: PreemptionRequested
    message: "Preemption on node 'node-1' succeeded in finding a suitable victim 'ns/pod-2' and successfully sent the victim removal request."
    lastTransitionTime: "2026-02-23T15:23:13Z"
```

The Scheduler also sets the condition to `False` with reason `ResizeFits` if the pod's target resource requirements fit on the node without requiring preemption. This signals that the scheduler believes that the node has sufficient capacity and is waiting for Kubelet to actuate the resize.

```yaml
status:
  conditions:
  - type: PodResizeUnschedulable
    status: "False"
    reason: ResizeFits
    message: "Node 'node-1' has sufficient capacity for the pod resize request."
    lastTransitionTime: "2026-02-23T15:23:13Z"
```

> The Scheduler clears this condition when the pod is no longer in the `Deferred` state or when the preemption policy is disabled (allowing the Kubelet to take ownership of the condition).

Reviewer 10: > or when the preemption policy is disabled

I'm not sure this would work well. That means the PRU condition would be temporarily removed after disabling preemption on the node? That requires the scheduler to watch the node preemption policy (which I thought we were trying to avoid), and also means the Kubelet needs to watch for the condition being removed after preemption is disabled, so that it can add it back?

Author: Changed the proposal to:
- PRU condition will stay (scheduler won't clear it)
- The scheduler will not watch the node preemption policy
- The kubelet will watch the node preemption policy and be responsible for updating the PRU condition in response to the disablement

##### 2. Kubelet-Owned (When Preemption is Disabled on the Node)
The Kubelet is the sole owner of the condition when preemption is disabled. If the Node's preemption policy prohibits preemption for resize requests, the Kubelet immediately sets `PodResizeUnschedulable` to `True` with reason `PreemptionDisabled` upon deferring the resize.

```yaml
status:
  conditions:
  - type: PodResizeUnschedulable
    status: "True"
    reason: PreemptionDisabled
    message: "Preemption for in-place pod resize is disabled on node 'node-1' by preemption policy."
    lastTransitionTime: "2026-02-23T15:23:13Z"
```

> Because the Scheduler will ignore deferred pods with the `PreemptionDisabled` reason, it will never write or overwrite this condition under this regime. The Kubelet clears this condition when the preemption policy is enabled (allowing Scheduler preemption to take over) or when the `Deferred` resize is resolved or cancelled.

Reviewer 10: This seems to conflict with the discussion of retries below?

Author: discussion of retries below was intended to only be applicable when the preemption policy is enabled. I updated it to state that explicitly

### Failures and Reconsideration of Deferred pods

#### PodResizeUnschedulable Pod Condition

To provide a unified view for external controllers, we propose a single new `PodResizeUnschedulable` condition type in the Pod status. This condition indicates that a `Deferred` resize request cannot proceed due to insufficient capacity and cannot benefit from scheduler-side preemption.

The condition is co-owned by the Scheduler and the Kubelet under mutually exclusive, deterministic ownership regimes based on whether preemption is enabled:

##### 1. Scheduler-Owned (When Preemption is Enabled on the Node)

The Scheduler is the sole owner of the condition when preemption is allowed. 

###### Reason: PreemptionFailed

The Scheduler sets the condition to `True` with reason `PreemptionFailed` if a preemption attempt fails to find suitable or eligible victims on the target node.

```yaml
status:
  conditions:
  - type: PodResizeUnschedulable
    status: "True"
    reason: PreemptionFailed
    message: "Preemption on node 'node-1' failed to find a suitable victim required for pod resize."
    lastTransitionTime: "2026-02-23T15:23:13Z"
```

> ###### Reason: PreemptionRequested

Author: I'm on the fence about whether we need to expose `PreemptionRequested` and `ResizeFits` cases (or just omit the condition when this happens); happy to either keep or remove them depending on feedback
